// @ericsson/oden/components/export/export-csv/export-csv.component.ts
import { Component, Input, OnInit } from '@angular/core';
import { CSVHeaderItem } from '../../../core/helpers/csv.helper';
import { ExportCSVService } from './export-csv.service';

@Component({
    selector: 'cui-export-csv',
    templateUrl: 'export-csv.component.html',
    styleUrls: ['export-csv.component.scss']
})

export class ExportCSVComponent implements OnInit {

    @Input() columns: Array<CSVHeaderItem> = [];
    selectedColumns: Array<CSVHeaderItem> = [];

    constructor(private _exportCSVService: ExportCSVService) {
    }

    ngOnInit(): void {
        this.columns = this.columns.filter(t => !t.hidden);
        this.selectedColumns = this.columns;
    }

    /**
    * Function to handle enable/disable for export button
    * @param {CSVHeaderItem[]} items
    */
    _onColumnChange(items: CSVHeaderItem[]): void {
        if (items.length >= 1) {
            this._exportCSVService._toggleExportDisable();
        } else {
            this._exportCSVService._toggleExportDisable(true);
        }
    }
}
