// @ericsson/cus-ui/dashboard/widget/widget-container/widget-set/widget-set.model.ts
import { QuickAssign } from '../../../helpers/quick-assign.model';

export class WidgetSet extends QuickAssign {
  private _data?: any[];
  currentData?: any[];
  private _dataSet?: any[];

  type: string;
  color?: string;
  title?: string;
  linkPath?: string;
  linkLabel?: string;
  unit?: string;
  trend?: string;
  trendLabel?: string;
  addons?: any[];
  xTitle?: string;
  yTitle?: string;
  sorted?: boolean;
  barChartType?: number;
  min?: number;
  max?: number;
  currentValue?: number;
  subtitle?: string;
  ticks?: number;
  tickFormat?: string;
  rotatedLabels?: boolean;
  colorSet?: string[];
  headline?: string;
  multiline?: boolean;
  columnSize?: string;
  groupKey?: string;
  externalApp?: string;
  columnMapping?: string;

  get data() {
    if (this.currentData) {
      return this.currentData;
    } else if (this.dataSet) {
      return this._dataSet[0].data;
    }
    return this._data;
  }

  set data(data: any[]) {
    this._data = data;
    this.currentData = data;
  }

  get dataSet() {
    return this._dataSet;
  }

  set dataSet(dataSet: any[]) {
    this._dataSet = dataSet;
  }
}
