// @ericsson/oden/components/dropzone/dropzone.module.ts
import { DropzoneService } from './dropzone.service';
import { DraggableModule } from './../draggable/draggable.module';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { DropzoneComponent } from './dropzone.component';
import { TactileBlockModule } from '../tactile-block/tactile-block.module';
import { CommonModule } from '@angular/common';
import { TactileBlockComponent } from '../tactile-block/tactile-block.component';
import { ContainerModule } from '../container/container.module';
import { IconModule } from '../icon/icon.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { ButtonModule } from '../button/button.module';
import { TranslateModule } from '@ngx-translate/core';
@NgModule({
    imports: [CommonModule, TactileBlockModule, DraggableModule, ContainerModule,
        IconModule, TooltipModule, ButtonModule, TranslateModule.forChild()],
    exports: [DropzoneComponent],
    declarations: [DropzoneComponent],
    entryComponents: [TactileBlockComponent]
})
export class DropzoneModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: DropzoneModule,
            providers: [DropzoneService]
        };
    }
}
