// @ericsson/oden/components/form/form.module.ts
import { TranslateModule } from '@ngx-translate/core';
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
import { TextMaskModule } from 'angular2-text-mask';
import { GestureService } from '../../core/services/gesture.service';
import { ButtonModule } from '../button/button.module';
import { ContainerModule } from '../container/container.module';
import { IconModule } from '../icon/icon.module';
import { CheckboxDirective } from './checkbox/checkbox.directive';
import { DateInputComponent } from './date-time/date-input.component';
import { DateTimeInputComponent } from './date-time/date-time-input.component';
import { TimeInputComponent } from './date-time/time-input.component';
import { DatePickerComponent } from './date-time-picker/date-picker/date-picker.component';
import { DateTimePickerComponent } from './date-time-picker/date-time-picker.component';
import { TimePickerComponent } from './date-time-picker/time-picker/time-picker.component';
import { DateTimeValidatorDirective } from './date-time/date-time.validators';
import { DateValidatorDirective } from './date-time/date.validators';
import { TimeValidatorDirective } from './date-time/time.validators';
import { InputDirective } from './input/input.directive';
import { InputSearchDirective } from './input/input-search.directive';
import { MultiButtonComponent } from './multi-button/multi-button.component';
import { PeriodPickerComponent } from './period-picker/period-picker.component';
import { ProgressBarComponent } from './progress-bar/progress-bar.component';
import { RadioDirective } from './radio/radio.directive';
import { RequiredDirective } from './required/required.directive';
import { SelectComponent } from './select/select.component';
import { OptionComponent } from './select/option/option.component';
import { DynamicSelectComponent } from './select/dynamic/dynamic-select.component';
import { SliderComponent } from './slider/slider.component';
import { SwitcherComponent } from './switcher/switcher.component';
import { TextareaDirective } from './textarea/textarea.directive';
import { UploadComponent } from './upload/upload.component';
import { PipesModule } from '../../core/pipes/pipes.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { OdenFormModule } from '../oden-form/oden-form.module';

@NgModule({
    imports: [CommonModule,
        IconModule,
        FormsModule,
        TextMaskModule,
        ContainerModule,
        ButtonModule,
        OdenFormModule,
        PipesModule,
        TooltipModule,
        TranslateModule.forChild()],
    exports: [
        CheckboxDirective,
        DateInputComponent,
        DatePickerComponent,
        DateTimeInputComponent,
        DateTimePickerComponent,
        InputDirective,
        InputSearchDirective,
        MultiButtonComponent,
        ProgressBarComponent,
        RadioDirective,
        RequiredDirective,
        TextareaDirective,
        TimeInputComponent,
        TimePickerComponent,
        SliderComponent,
        SwitcherComponent,
        SelectComponent,
        OptionComponent,
        UploadComponent,
        DateValidatorDirective,
        DateTimeValidatorDirective,
        TimeValidatorDirective,
        DynamicSelectComponent,
        PeriodPickerComponent
    ],
    declarations: [
        CheckboxDirective,
        DateInputComponent,
        DatePickerComponent,
        DateTimeInputComponent,
        DateTimePickerComponent,
        InputDirective,
        InputSearchDirective,
        MultiButtonComponent,
        ProgressBarComponent,
        RadioDirective,
        RequiredDirective,
        TextareaDirective,
        SliderComponent,
        SwitcherComponent,
        SelectComponent,
        TimeInputComponent,
        TimePickerComponent,
        OptionComponent,
        UploadComponent,
        DateValidatorDirective,
        DateTimeValidatorDirective,
        TimeValidatorDirective,
        DateTimeValidatorDirective,
        DynamicSelectComponent,
        PeriodPickerComponent
    ]
})

export class FormModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: FormModule,
            providers: [{provide: HAMMER_GESTURE_CONFIG, useClass: GestureService}]
        };
    }
}
