// @ericsson/oden/components/table2/head/head-title/head-count/count.component.ts
import {
    ChangeDetectionStrategy,
    ChangeDetectorRef,
    Component,
    forwardRef,
    Inject,
    Input,
    OnChanges,
    OnInit,
    SimpleChanges
} from '@angular/core';
import { Table2Component } from '../../../table.component';
import { Subscription } from 'rxjs';
import { UnSub } from '../../../../../core/classes/subscription.decorator';

@Component({
    selector: 'oden-table-title-count',
    templateUrl: './count.component.html',
    styleUrls: ['./count.component.scss'],
    changeDetection: ChangeDetectionStrategy.OnPush
})
@UnSub()
export class TableTitleCountComponent implements OnInit, OnChanges {
    @Input()
    count: number;
    _displayCount: number;
    private _subscriptions: Array<Subscription> = [];
    constructor(
        @Inject(forwardRef(() => Table2Component))
        private _table: Table2Component,
        private _changeDetectorRef: ChangeDetectorRef
    ) {}
    ngOnInit() {
        this.setCount();
        this._table.tableSetupCompleted.first().subscribe(() => {
            this.setCount();
            this._subscriptions.push(
                this._table.rowComponentsQueryList.changes.subscribe(() =>
                    this.setCount()
                )
            );
        });
    }

    setCount(): void {
        if (typeof this.count === 'undefined') {
            this._displayCount = this._table.rowComponents.length;
        } else {
            this._displayCount = this.count;
        }
        this._changeDetectorRef.detectChanges();
    }
    ngOnChanges(changes: SimpleChanges) {
        if (changes['count']) {
            this.setCount();
        }
    }
}
