// @ericsson/oden/components/app-header/app-header.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';

import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import { ButtonModule } from '../button/button.module';
import { FormModule } from '../form/form.module';
import { IconModule } from '../icon/icon.module';
import { StickyModule } from '../sticky/sticky.module';
import { TabsModule } from '../tabs/tabs.module';
import { AppHeaderComponent } from './app-header.component';
import { TooltipModule } from '../tooltip/tooltip.module';

@NgModule({
    imports: [
        BreadcrumbModule,
        ButtonModule,
        CommonModule,
        TabsModule,
        StickyModule,
        IconModule,
        FormModule,
        FormsModule,
        TranslateModule.forChild(),
        TooltipModule
    ],
    exports: [AppHeaderComponent],
    declarations: [AppHeaderComponent]
})

export class AppHeaderModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: AppHeaderModule
        };
    }
}
