// @ericsson/oden/components/chart/simple/simple.component.ts
import {
    Component,
    ComponentFactoryResolver,
    HostBinding,
    Input, OnInit,
    ViewEncapsulation
} from '@angular/core';
import { ChartData } from '../../../core/models';
import { ChartService } from '../chart.service';
import { ChartWithAddons } from '../chart-with-addons.class';

@Component({
    selector: 'cui-simple',
    templateUrl: 'simple.component.html',
    styleUrls: ['simple.component.scss'],
    providers: [ChartService],
    encapsulation: ViewEncapsulation.None
})

export class SimpleComponent extends ChartWithAddons implements OnInit {
    @HostBinding('class.cui-simple') addClass: boolean = true;
    @Input() dataSet: Array<ChartData> = [];
    @Input() title: string;
    @Input() color: string;

    constructor(componentFactoryResolver: ComponentFactoryResolver,
                private _chartService: ChartService) {
        super(componentFactoryResolver, _chartService);
    }

    ngOnInit(): void {
        this.reload();
    }

    /**
     * Force reload when data changed
     */
    reload(): void {
        this._chartService.color = this.color;
        this._chartService.data = this.dataSet;
        this._chartService.isInteractive = false;
        this.createAddons(this.addonsContainer, this.addons);
    }
}
