// @ericsson/oden/components/characteristics/characteristics.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';

import { ContainerModule } from '../container/container.module';
import { BlockModule } from '../block/block.module';
import { ListViewModule } from '../list-view/list-view.module';
import { PaginationModule } from '../pagination/pagination.module';
import { FormModule } from '../form/form.module';
import { TooltipModule } from '../tooltip/tooltip.module';

import { CharacteristicsComponent } from './characteristics.component';
import { CharacteristicHelperService } from './characteristic.helper.service';
import { CharacteristicContainerComponent } from './characteristic-container/characteristic-container.component';
import { CharacteristicBlockComponent } from './characteristic-container/characteristic-block/characteristic-block.component';
import { StringTypeComponent } from './characteristic-container/characteristic-types/char-type-string/string.component';
import { LongTypeComponent } from './characteristic-container/characteristic-types/char-type-long/long.component';
import { DecimalTypeComponent } from './characteristic-container/characteristic-types/char-type-decimal/decimal.component';
import { BooleanTypeComponent } from './characteristic-container/characteristic-types/char-type-boolean/boolean.component';
import { DateTypeComponent } from './characteristic-container/characteristic-types/char-type-date/date.component';
import { TimeTypeComponent } from './characteristic-container/characteristic-types/char-type-time/time.component';
import { DateTimeTypeComponent } from './characteristic-container/characteristic-types/char-type-date-time/date-time.component';
import { DateTimePeriodTypeComponent }
    from './characteristic-container/characteristic-types/char-type-date-time-period/date-time-period.component';
import { TimeOfDayPeriodTypeComponent }
    from './characteristic-container/characteristic-types/char-type-time-of-day-period/time-of-day-period.component';
import { ElementOverlayModule } from '../element-overlay/element-overlay.module';

@NgModule({
    imports: [
        CommonModule,
        ContainerModule,
        BlockModule,
        ListViewModule,
        PaginationModule,
        FormsModule,
        FormModule,
        TooltipModule,
        ElementOverlayModule
    ],
    exports: [CharacteristicsComponent],
    declarations: [
        CharacteristicsComponent,
        CharacteristicContainerComponent,
        CharacteristicBlockComponent,
        StringTypeComponent,
        LongTypeComponent,
        DecimalTypeComponent,
        BooleanTypeComponent,
        DateTypeComponent,
        TimeTypeComponent,
        DateTimeTypeComponent,
        DateTimePeriodTypeComponent,
        TimeOfDayPeriodTypeComponent
    ],
    entryComponents: [
        StringTypeComponent,
        LongTypeComponent,
        DecimalTypeComponent,
        BooleanTypeComponent,
        DateTypeComponent,
        TimeTypeComponent,
        DateTimeTypeComponent,
        DateTimePeriodTypeComponent,
        TimeOfDayPeriodTypeComponent
    ],
    providers: [CharacteristicHelperService]
})

export class CharacteristicsModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: CharacteristicsModule
        };
    }
}
