// @ericsson/oden/components/drawer/drawer.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';

import { ContainerModule } from '../container/container.module';
import { IconModule } from '../icon/icon.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { DrawerComponent } from './drawer.component';
import { DrawerService } from './drawer.service';
import { MultiDrawerComponent } from './multi-drawer/multi-drawer.component';
import { AppHeaderModule } from '../app-header/app-header.module';

@NgModule({
    imports: [CommonModule, ContainerModule, IconModule, TooltipModule, AppHeaderModule],
    exports: [DrawerComponent, MultiDrawerComponent],
    declarations: [DrawerComponent, MultiDrawerComponent],
    entryComponents: [DrawerComponent, MultiDrawerComponent]

})

export class DrawerModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: DrawerModule,
            providers: [DrawerService]
        };
    }
}
