// @ericsson/oden/components/list-view/list-view.module.ts
import { CommonModule } from '@angular/common';
import { NgModule, ModuleWithProviders } from '@angular/core';
import { ContainerModule } from '../container/container.module';
import { ListViewComponent } from './list-view.component';
import { ListViewHeadlineComponent } from './list-view-headline/list-view-headline';
import { ListViewSubheadlineComponent } from './list-view-subheadline/list-view-subheadline';
import { ListViewLabelComponent } from './list-view-label/list-view-label';
import { ListViewValueComponent } from './list-view-value/list-view-value';
import { TooltipModule } from '../tooltip/tooltip.module';
import { FormModule } from '../form/form.module';

@NgModule({
    imports: [CommonModule, ContainerModule, TooltipModule, FormModule],
    exports: [
        ListViewComponent,
        ListViewHeadlineComponent,
        ListViewSubheadlineComponent,
        ListViewLabelComponent,
        ListViewValueComponent
    ],
    declarations: [
        ListViewComponent,
        ListViewHeadlineComponent,
        ListViewSubheadlineComponent,
        ListViewLabelComponent,
        ListViewValueComponent
    ]
})

export class ListViewModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: ListViewModule
        };
    }
}
