// @ericsson/oden/components/relation-graph/graph/constants.ts
export const RelationGraphComponentConstants = {
    TRANSITION: {
        DURATION: 250
    },
    DEPTH_CONTROLLER: {
        WIDTH: '100%',
        HEIGHT: 32,
        DEPTH_CONTROL: {
            RADIUS: 10,
            PADDING: 6
        }
    },
    GRID: {
        SIZE: {
            WIDTH: 200,
            HEIGHT: 50
        },
        SCALE: {
            MIN: 0.6,
            MAX: 1.4,
            STEP: 0.1
        }
    },
    FOCUS: {
        DELAY: 500,
        TIMEOUT: 5000
    },
    FILTER: {
        DELAY: 100
    },
    NODE: {
        WIDTH: 48,
        HEIGHT: 24,
        NAME: {
            MARGIN: 2
        },
        LINK_AWAY: {
            RADIUS: 8,
            ARROW: 3,
            DX: -4,
            DY: 2
        },
        TAG: {
            HEIGHT: 12,
            PADDING_LEFT: 4,
            PADDING_RIGHT: 2,
            ALIGN: 0,
            DX: -16
        }
    },
    ROOT_NODE: {
        WIDTH: 172,
        HEIGHT: 24,
        PADDING: 24,
        TAG: {
            ALIGN: -1,
            DX: -12
        }
    }
};
