// @ericsson/oden/components/table/table.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';

import { ButtonModule } from '../button';
import { ContainerModule } from '../container/container.module';
import { FormModule } from '../form';
import { IconModule } from '../icon/icon.module';
import { ListControllerModule } from '../list-controller';
import { LoaderModule } from '../loader';
import { PaginationModule } from '../pagination';
import { TooltipModule } from '../tooltip/tooltip.module';
import { TableRowModule } from './table-row/table-row.module';
import { TableComponent } from './table.component';
import { TableService } from './table.service';
import { TableSortingDirective } from './table-sorting/table-sorting.directive';
import { TranslateModule } from '@ngx-translate/core';
import { BatchActionsModule } from '../batch-actions';
import { ActionMenuModule } from '../action-menu/action-menu.module';
import { ExportModule } from '../export/export.module';

@NgModule({
    imports: [CommonModule, ButtonModule, FormModule, PaginationModule, ContainerModule, FormsModule,
        LoaderModule, IconModule, TableRowModule, ListControllerModule, TooltipModule, TranslateModule.forChild(),
    BatchActionsModule, ActionMenuModule, ExportModule],
    exports: [TableComponent, TableSortingDirective],
    declarations: [TableComponent, TableSortingDirective]
})

export class TableModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: TableModule,
            providers: [TableService]
        };
    }
}
