// @ericsson/oden/components/action-menu/action-menu.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';

import { IconModule } from '../icon/icon.module';
import { ActionMenuItemComponent } from './action-menu-item';
import { ActionMenuComponent } from './action-menu.component';
import { PipesModule } from '../../core/pipes/pipes.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { BackdropModule } from '../backdrop';
import { TranslateModule } from '@ngx-translate/core';
import { PreventDraggingModule } from '../prevent-dragging/prevent-dragging.module';

@NgModule({
    imports: [CommonModule, IconModule, PipesModule, TooltipModule, BackdropModule, TranslateModule.forChild(), PreventDraggingModule],
    exports: [ActionMenuComponent, ActionMenuItemComponent],
    declarations: [ActionMenuComponent, ActionMenuItemComponent],
})

export class ActionMenuModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: ActionMenuModule
        };
    }
}
