// @ericsson/oden/components/overlay/overlayable.class.ts
import { OnDestroy, ElementRef, ComponentRef, Renderer2 } from '@angular/core';
import { BackdropComponent } from '../backdrop/backdrop.component';
import { OverlayHelperService } from './overlay-helper.service';
import { IOverlayable } from './overlayable.interface';

export class Overlayable implements OnDestroy, IOverlayable {
    renderer: Renderer2;
    overlayElement: ElementRef;
    public elementRef: ElementRef;
    public overlayHelperService: OverlayHelperService;
    openedInOverlay: boolean = false;
    backdrop: ComponentRef<BackdropComponent>;

    ngOnDestroy(): void {
        this.removeFromOverlay();
    }

    /**
     * Call this when ever you want to remove the overlayElement - e.g on scroll.
     */
    removeFromOverlay(): void {
        if (this.overlayElement && this.openedInOverlay) {
            this.overlayHelperService.removeFromOverlay(this);
        }
    }

    /**
     * Call this function after you have set the elementRef and overlayElement
     */
    openInOverlay(): void {
        this.overlayHelperService.appendToOverlay(this);
    }
}
