// @ericsson/oden/core/services/library.service.ts
import { EventEmitter, Injectable } from '@angular/core';
import { LibraryAction } from '../models/library-action.model';
import { PaginationData } from '../../components/pagination/pagination.model';

@Injectable()
export class LibraryService {

    onFilter: EventEmitter<string> = new EventEmitter<string>();
    onMenuActionClick: EventEmitter<LibraryAction> = new EventEmitter<LibraryAction>();
    onButtonActionClick: EventEmitter<LibraryAction> = new EventEmitter<LibraryAction>();
    onCategoryFilterChange: EventEmitter<any> = new EventEmitter<any>();
    public added$: EventEmitter<any> = new EventEmitter<any>();
    onPagination: EventEmitter<PaginationData> = new EventEmitter<PaginationData>();

    /**
     * @param obj Consists of 'group' and 'ngModel'
     */
    categoryFilterChange(obj: any): void {
        this.onCategoryFilterChange.emit(obj);
    }

    buttonActionClick(libraryAction: LibraryAction) {
        this.onButtonActionClick.emit(libraryAction);
    }

    menuActionClick(libraryAction: LibraryAction): void {
        this.onMenuActionClick.emit(libraryAction);
    }

    filterChange(string: string): void {
        this.onFilter.emit(string);
    }

    paginatonClick(paginationData: PaginationData) {
        this.onPagination.emit(paginationData);
    }

}
