// @ericsson/oden/components/chart/addon/selector/selector-addon.component.ts
import { Component, HostBinding, Input, OnInit, ViewEncapsulation } from '@angular/core';
import { OptionItem } from '../../../../core/models/option-item.model';
import { UUID } from '../../../../core/helpers/uuid.class';

@Component({
    selector: 'cui-selector-addon',
    templateUrl: 'selector-addon.component.html',
    styleUrls: ['selector-addon.component.scss'],
    encapsulation: ViewEncapsulation.None
})

export class SelectorAddonComponent implements OnInit {
    @HostBinding('class.cui-selector-addon') addClass: boolean = true;
    @Input() options: Array<OptionItem> = [];
    @Input() onChange;
    @Input() label: string = '';
    @Input() id: string = 'select-addon-' + UUID.create();
    activeItem: OptionItem = null;

    ngOnInit() {
        this.activeItem = this.options.find(option => option.selected);
    }

    onSelect(event) {
        if (this.onChange && typeof this.onChange === 'function') {
            this.onChange(event);
        }
        this.activeItem = event;
    }
}
