// @ericsson/oden/components/tabs/ink-bar/inkbar.component.ts
import { Component, Renderer2, ElementRef, HostBinding, ViewEncapsulation } from '@angular/core';
import { DOMHelper } from '../../../core/helpers/dom.helper';

/** The ink-bar is used to display and animate the line underneath the current active tab label. */
@Component({
    selector: 'cui-inkbar',
    template: '',
    styleUrls: ['inkbar.component.scss'],
    encapsulation: ViewEncapsulation.None
})

export class InkBarComponent {

    @HostBinding('class.cui-inkbar') addClass: boolean = true;
    constructor(private _renderer: Renderer2,
                public elementRef: ElementRef) {}

    /**
     * Calculates the styles from the provided element in order to align the ink-bar to that element.
     * @param element
     * @param shrinkPx - Amount of pixels to shrink inkbar on left and right
     */
    public alignToElement(element: HTMLElement, shrinkPx: number = 0) {
        this._renderer.setStyle(this.elementRef.nativeElement, 'left',
            DOMHelper.getLeftPosition(element) + 'px');
        this._renderer.setStyle(this.elementRef.nativeElement, 'width',
            (DOMHelper.getOuterWidth(element) - (shrinkPx * 2)).toString() + 'px');
        this._renderer.setStyle(this.elementRef.nativeElement, 'marginLeft', shrinkPx + 'px');
    }
}
