// @ericsson/cus-ui/task/mocks/engine/utils.js
var fs = require('fs');

function notEmpty(input) {
    return !!input;
}

function byId(objs, id) {
    var keys = Object.keys(objs);
    for (var k in keys) {
        if (objs[keys[k]].id === id) {
            return objs[keys[k]];
        }
    }
}

function toArray(obj) {
    var arr = [];
    Object.keys(obj).forEach(function (key) {
        arr.push(obj[key]);
    });
    return arr;
}

function thr(err) {
    if (err) {
        throw err;
    }
}

// returns an array of functions reading the content of files.
// To be used in CamSDK.utils.series()
function readFiles(dirPath, filenames) {
    return filenames.map(function (filename) {
        return function (cb) {
            fs.readFile(path.join(dirPath, filename), function (err, content) {
                if (err) { return cb(err); }

                cb(null, {
                    name: filename,
                    content: content.toString()
                });
            });
        };
    });
}

module.exports = {
    notEmpty: notEmpty,
    byId: byId,
    toArray: toArray,
    thr: thr,
    readFiles: readFiles
}
