// @ericsson/oden/components/form/select/select.service.ts

import {
    Injectable, EventEmitter
} from '@angular/core';
import { OptionComponent } from './option/option.component';

@Injectable()
export class SelectService {

    onSelect: EventEmitter<OptionComponent> = new EventEmitter<OptionComponent>();
    onMultiSelect: EventEmitter<OptionComponent> = new EventEmitter<OptionComponent>();
    onSelectAll: EventEmitter<any> = new EventEmitter<any>();

    noDataMessage: string = 'No data found';
    filterMessage: string = 'Please try another filtering criteria';
    dynamicMessage: string = 'Add your new value with Add button above';
    noMatchMessage: string = 'No matching results';

    value: any;
    multiple: boolean;
    origin: HTMLElement;

    /**
     * Sends the param option to either selectMultipleOption or selectSingleOption,
     * depending on value of bool multiple
     * @param option
     */
    setOption(option: OptionComponent): void {
        if (this.multiple) {
           this.onMultiSelect.emit(option);
        } else {
            this.onSelect.emit(option);
        }
    }

    setOrigin(element: HTMLElement): void {
        this.origin = element;
    }

    toggleSelectAll() {
        this.onSelectAll.emit();
    }

}
