// @ericsson/cus-ui/task/management/models/task-list.model.ts
import { TableHeaderItem } from '@ericsson/oden';
import {
    ListControllerCategoryFilterGroup,
    ListControllerCategoryFilterItem,
    StateBatch,
    TableRowData,
} from '@ericsson/oden/core';

export class AdditionalData {
    key: string;
    value: string;
}

export class TaskListItem extends TableRowData {
    serialNo: number;
    id: string;
    state: StateBatch;
    type: string;
    name: string;
    taskDisplayName: string;
    instanceName: string;
    workflowName: string;
    instanceId: string;
    workflowId: string;
    description: string;
    assignee: string;
    screenPath: string;
    owner: string;
    dueDate: Date;
    timeLeft: string;
    assignUrl: string;
    tpgName: string;
    revokeUrl: string;
    listUrl: string;
    customerName: string;
    additionalData: Array<AdditionalData>;
    task?: TaskListItem[];
    searchMatch?: boolean;
    isExpanded?: boolean;
}

export class TaskList {
    tasks: Array<TaskListItem>;
}

export class TaskOverview extends TableRowData {
    serialNo: number;
    taskId: string;
    taskName: string;
    description: string;
}

export class CategoryFilter {
    group: ListControllerCategoryFilterGroup;
    ngModel: ListControllerCategoryFilterItem[];
}

export class GroupByFilterData {
    data: TaskListItem[];
    header: TableHeaderItem[];
    name?: string;
    isPagination: boolean;
}
