// @ericsson/oden/components/table/table-row/table-cell/table-cell-edit/table-cell-edit-text/table-cell-edit-text.component.ts
import { Component, Input, OnChanges, OnInit, ViewEncapsulation } from '@angular/core';
import { TableCellEditComponent } from '../table-cell-edit.component';
import { TABLE_CELL_EDIT } from '../table-cell-edit.animations';

@Component({
    selector: 'cui-table-cell-edit-text',
    templateUrl: './table-cell-edit-text.component.html',
    encapsulation: ViewEncapsulation.None,
    animations: [ TABLE_CELL_EDIT.ANIMATION ]
})

export class TableCellEditTextComponent extends TableCellEditComponent implements OnInit, OnChanges {
    @Input() editData: string = null;
    @Input() editOptions: {strictPattern?: RegExp, maxLength?: number} = {};
    private _previousValue = '';
    private _previousEditData;

    ngOnInit(): void {
        this.editOptions = this.editOptions || {};
        this._setValues();
    }

    ngOnChanges(): void {
        this._setValues();
    }

    /**
     * Listens for change in the input.
     * @param value
     * @param element
     */
    onChange(value: string, element: HTMLInputElement): void {
        super.onChange(value, element);
    }

    /**
     * Sets values for editData and values if it has been changed.
     * @private
     */
    private _setValues(): void {
        if (this._previousEditData !== this.editData) {
            this._previousEditData = this._previousValue = this.editData;
        }
    }
}
