// @ericsson/oden/components/table2/table2-row/single-select-cell/single-select-cell.component.ts
import {
    Component,
    Input,
    OnInit,
    Inject,
    forwardRef,
    EventEmitter,
    Output,
    ChangeDetectorRef
} from '@angular/core';
import { Row2Component } from '../row.component';
import { Table2Component } from '../../table.component';

@Component({
    // tslint:disable-next-line:component-selector
    selector: '[odenSingleSelectCell]',
    templateUrl: './single-select-cell.component.html',
    host: { class: 'select' }
})
export class TableSingleSelectCellComponent implements OnInit {
    constructor(
        @Inject(forwardRef(() => Row2Component)) private _row: Row2Component,
        @Inject(forwardRef(() => Table2Component)) public _table: Table2Component,
        private _cdr: ChangeDetectorRef
    ) {}
    _selected = false;
    @Input()
    set selected(value: boolean) {
        this._selected = value;
        this._cdr.detectChanges();
    }
    get selected(): boolean {
        return this._selected;
    }

    @Output()
    onChange = new EventEmitter<boolean>(true);
    _rowUuid: string;
    ngOnInit(): void {
        this._rowUuid = this._row._rowUuid;
        this._row._selected = this.selected;
    }
    toggleSelectValue() {
        this.selected = !this.selected;
        this.onChange.emit(this.selected);
    }
}
