// @ericsson/oden/components/list-controller/list-controller.module.ts
import { TogglerModule } from './../toggler/toggler.module';
import { TranslateModule } from '@ngx-translate/core';
import { CommonModule } from '@angular/common';
import { NgModule, ModuleWithProviders } from '@angular/core';
import { FormsModule } from '@angular/forms';

import { ListControllerComponent } from './list-controller.component';
import { ListControllerOrderComponent } from './list-controller-order';
import { ListControllerFilterComponent } from './list-controller-filter';
import { ListControllerActionComponent } from './list-controller-action';
import { ActionMenuModule } from '../action-menu';
import { FormModule } from '../form';
import { IconModule } from '../icon/icon.module';
import { ListControllerCategoryFilterComponent } from './list-controller-category-filter/list-controller-category-filter.component';
import { ListControllerGeneralComponent } from './list-controller-general/list-controller-general.component';
import { ListControllerNumberOfResultsComponent } from './list-controller-number-of-results/list-controller-number-of-results.component';
import { PipesModule } from '../../core/pipes/pipes.module';
import { ListControllerClearFiltersComponent } from './list-controller-clear-filters/list-controller-clear-filters.component';
import { TooltipModule } from '../tooltip/tooltip.module';
import { ModalModule } from '../modal';
import { ContainerModule } from '../container/container.module';
import { LoaderModule } from '../loader/loader.module';

@NgModule({
    imports: [
        CommonModule,
        ActionMenuModule,
        FormsModule,
        FormModule,
        IconModule,
        PipesModule,
        TooltipModule,
        ModalModule,
        TogglerModule,
        ContainerModule,
        LoaderModule,
        TranslateModule.forChild()
    ],
    exports: [
        ListControllerComponent,
        ListControllerActionComponent,
        ListControllerFilterComponent,
        ListControllerCategoryFilterComponent,
        ListControllerOrderComponent,
        ListControllerGeneralComponent,
        ListControllerNumberOfResultsComponent,
        ListControllerClearFiltersComponent
    ],
    declarations: [
        ListControllerComponent,
        ListControllerActionComponent,
        ListControllerFilterComponent,
        ListControllerCategoryFilterComponent,
        ListControllerOrderComponent,
        ListControllerGeneralComponent,
        ListControllerNumberOfResultsComponent,
        ListControllerClearFiltersComponent
    ]
})
export class ListControllerModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: ListControllerModule
        };
    }
}
