// @ericsson/oden/components/node-tree/node-tree.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { NodeTreeComponent } from './node-tree.component';
import { NodeViewComponent } from './node-view/node-view.component';
import { IconModule } from '../icon/icon.module';
import { FormsModule } from '@angular/forms';
import { ListControllerModule } from '../list-controller/list-controller.module';
import { PromoNodeTreeComponent } from './promo-node-tree/promo-node-tree.component';
import { PromoNodeViewComponent } from './promo-node-tree/promo-node-view/promo-node-view.component';
import { TooltipModule } from '../../components/tooltip/tooltip.module';

@NgModule({
    imports: [CommonModule, IconModule, FormsModule, ListControllerModule, TooltipModule],
    exports: [NodeTreeComponent],
    declarations: [NodeTreeComponent, NodeViewComponent, PromoNodeTreeComponent, PromoNodeViewComponent]
})

export class NodeTreeModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: NodeTreeModule
        };
    }
}
