// @ericsson/oden/core/interfaces/table-header.interface.ts
import { ORDER, ALIGNMENT, EDIT_TYPE } from '../constants';
import { ORDER_TYPE } from '../constants/order-type.constants';
import { FIELD_TYPE } from '../constants/field-type.constants';

export interface ITableHeaderItem {
    label: string;
    order?: ORDER;
    field: string;
    editType?: EDIT_TYPE;
    /** @deprecated use templates! */
    editField?: string;
    /** @deprecated use templates! */
    fieldType?: FIELD_TYPE;
    orderType?: ORDER_TYPE;
    orderField?: string;
    icon?: string;
    sortOrders?: ORDER[]; // Limits sorting to specific set of sort orders
    tooltip?: string; // Shows up if you hover the header - currently unused.
    align?: ALIGNMENT; // Align Left, right och center
    width?: string; // Sets width of the column. in pixels or percent. for example: '150px' or '25%'
    hidden?: boolean; // Sets hidden state for the column
    overflowed?: boolean; // Set internally by table component
    sticky?: boolean; // If this column should always be visible
    /** @deprecated use templates! */
    editData?: any;
    /** @deprecated use templates! */
    editRequired?: boolean;
    /** @deprecated use templates! */
    editPlaceholder?: string;


}
