// @ericsson/oden/components/node-tree/promo-node-tree/promo-node-view/promo-node-view.component.ts
/* tslint:disable:component-selector */
import {
    Component,
    HostBinding,
    OnDestroy,
    OnInit,
    ViewEncapsulation,
    HostListener,
    TemplateRef
} from '@angular/core';
import { PromoNodeTreeService } from '../promo-node-tree.service';
import { NodeViewComponent } from '../../node-view/node-view.component';

@Component({
    selector: '[cui-promo-node-view]',
    templateUrl: './promo-node-view.component.html',
    styleUrls: ['./promo-node-view.component.scss'],
    encapsulation: ViewEncapsulation.None
})

export class PromoNodeViewComponent extends NodeViewComponent implements OnInit, OnDestroy {
    @HostBinding('class.cui-node-view') add: boolean = false;
    @HostBinding('class.cui-promo-node-view') addClass: boolean = true;
    @HostBinding('class.active') get isActive() {
        return this.node.active;
    }
    @HostBinding('class.active-parent') get isActiveParent() {
        return this.node.activeParent;
    }
    @HostBinding('class.custom-template') get hasCustomTemplate() {
        return this.node.template;
    }
    @HostBinding('class.template-reference') hasTemplateRef: boolean;

    @HostListener('click') get addEventListener() {
        return this.toggleActive;
    }

    constructor(private promoNodeTreeService: PromoNodeTreeService) {
        super(promoNodeTreeService);
    }

    ngOnInit(): void {
        super.ngOnInit();
        this.node.icon = this.node.icon || 'favorite-unselected'; // icon is mandatory
        if (!this.node.active && this.promoNodeTreeService.isAnyChildNodeActive(this.node)) {
            this.node.activeParent = true;
        }
        this.hasCustomTemplateOrComponent();
    }
    /**
     * Identifies if custom template or component is available
     */
    hasCustomTemplateOrComponent() {
        if (this.node.template) {
            this.hasTemplateRef =  this.node.template instanceof TemplateRef ? true : false;
        }
    }
}
