// @ericsson/cus-ui/common/common.module.ts
import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule as AngularCommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { OdenModule } from '@ericsson/oden/modules';
import { I18nModule } from '../i18n';
import { LoggerModule } from '../logger';
import { ErrorComponent, TableErrorComponent, UniversalOverlayComponent } from './components';
import { AuthenticationModule } from '../core/authentication/authentication.module';

import {
    ApplicationSettings,
    EventService,
    ExceptionHandlerService,
    FunctionControlService,
    HttpInterceptor,
    MainMenuService,
    MetaDataService,
    UserProfileService,
    ErrorPageService,
    DialogService,
    RoutingHistoryService,
    ErrorHandlerService
} from './services';

const MODULE_IMPORTS = [
    AngularCommonModule,
    HttpClientModule,
    OdenModule,
    I18nModule,
    LoggerModule,
    AuthenticationModule
]

const MODULE_PROVIDERS = [
    ApplicationSettings,
    EventService,
    ExceptionHandlerService,
    FunctionControlService,
    HttpInterceptor,
    MainMenuService,
    MetaDataService,
    UserProfileService,
    ErrorPageService,
    DialogService,
    RoutingHistoryService,
    ErrorHandlerService
]

const MODULE_DECLARATION = [
    ErrorComponent,
    TableErrorComponent,
    UniversalOverlayComponent
]

@NgModule({
    imports: MODULE_IMPORTS,
    providers: [],
    declarations: MODULE_DECLARATION,
    exports: [TableErrorComponent, UniversalOverlayComponent],
    entryComponents: []
})

export class CommonModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: CommonModule,
            providers: MODULE_PROVIDERS
        };
    }
}
