// @ericsson/oden/components/rule-container/rule-container.module.ts
import { ModuleWithProviders, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RuleContainerComponent } from './rule-container.component';
import { RuleActionComponent } from './rule-action/rule-action.component';
import { RuleTreeComponent } from './rule-tree/rule-tree.component';
import { RuleConditionComponent } from './rule-condition/rule-condition.component';
import { IconModule } from '../icon';
import { RuleEventComponent } from './rule-event/rule-event.component';
import { TooltipModule } from '../tooltip';
import { PipesModule } from '../../core/pipes/pipes.module';
import { TranslateModule } from '@ngx-translate/core';

@NgModule({
    imports: [IconModule, CommonModule, TooltipModule, PipesModule, TranslateModule.forChild()],
    exports: [RuleContainerComponent, RuleActionComponent, RuleEventComponent, RuleConditionComponent, RuleTreeComponent],
    declarations: [
        RuleContainerComponent, RuleActionComponent, RuleEventComponent,
        RuleConditionComponent, RuleTreeComponent
    ]
})
export class RuleContainerModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: RuleContainerModule
        };
    }
}
