// @ericsson/oden/components/all/oden.module.ts
import { OdenCheckboxDirective } from './../oden-form/oden-checkbox/oden-checkbox.directive';
import { OdenFormModule } from './../oden-form/oden-form.module';
import { TogglerExpandComponent } from './../toggler/toggler-expand/toggler-expand.component';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { ApplicationService, PostMessageService, ResizeService, WhitelabelService } from '../../core';
import { FilterService } from '../../core/services/filter.service';
import { VersionsService } from '../../core/services/versions.service';
import { ActionMenuModule } from '../action-menu';
import { ActionMenuItemComponent } from '../action-menu/action-menu-item/action-menu-item';
import { ActionMenuComponent } from '../action-menu/action-menu.component';
import {
    AppHeaderComponent,
    AppHeaderModule,
    AppHeaderService
} from '../app-header';
import { BackdropComponent, BackdropModule } from '../backdrop';
import { BlockComponent, BlockModule } from '../block';
import {
    BlockTreeComponent,
    BlockTreeBranchComponent,
    BlockTreeBranchLeafComponent,
    BlockTreeModule
} from './../block-tree';
import { BreadcrumbModule } from '../breadcrumb';
import {
    ButtonComponent,
    ButtonModule,
    ButtonWrapperDirective
} from '../button';
import { CallOutComponent, CallOutModule } from '../call-out';
import { AddonModule } from '../chart/addon/addon.module';
import { BarChartComponent } from '../chart/bar-chart/bar-chart.component';
import { BarChartModule } from '../chart/bar-chart/bar-chart.module';
import { GaugeChartComponent } from '../chart/gauge-chart/gauge-chart.component';
import { GaugeChartModule } from '../chart/gauge-chart/gauge-chart.module';
import { LineChartComponent, LineChartModule } from '../chart/line-chart';
import { PieChartComponent } from '../chart/pie-chart/pie-chart.component';
import { PieModule } from '../chart/pie-chart/pie-chart.module';
import { SimpleComponent } from '../chart/simple/simple.component';
import { SimpleModule } from '../chart/simple/simple.module';
import {
    ColumnComponent,
    ColumnSizeDirective,
    ContainerComponent,
    RowComponent
} from '../container';
import { ContainerModule } from '../container/container.module';
import { DraggableDirective, AutoScrollDirective } from '../draggable';
import { DraggableModule } from '../draggable/draggable.module';
import { DrawerComponent, DrawerModule, MultiDrawerComponent } from '../drawer';
import { DropzoneComponent, DropzoneModule } from '../dropzone';
import { ElementOverlayDirective } from '../element-overlay/element-overlay.directive';
import { ElementOverlayModule } from '../element-overlay/element-overlay.module';
import { EntityVersionComponent, EntityVersionModule } from '../entity-version';
import { ErrorPageComponent, ErrorPageModule } from '../error-page/';
import { CheckboxDirective } from '../form/checkbox/checkbox.directive';
import {
    DateInputComponent,
    DateTimeInputComponent,
    TimeInputComponent
} from '../form/date-time';
import {
    DatePickerComponent,
    DateTimePickerComponent,
    TimePickerComponent
} from '../form/date-time-picker';
import { DateTimeValidatorDirective } from '../form/date-time/date-time.validators';
import { DateValidatorDirective } from '../form/date-time/date.validators';
import { TimeValidatorDirective } from '../form/date-time/time.validators';
import { FormModule } from '../form/form.module';
import { InputDirective, InputSearchDirective } from '../form/input';
import { MultiButtonComponent } from '../form/multi-button/multi-button.component';
import { PeriodPickerComponent } from '../form/period-picker/period-picker.component';
import { ProgressBarComponent } from '../form/progress-bar';
import { RadioDirective } from '../form/radio/radio.directive';
import { RequiredDirective } from '../form/required/required.directive';
import {
    DynamicSelectComponent,
    OptionComponent,
    SelectComponent
} from '../form/select';
import { SliderComponent } from '../form/slider/slider.component';
import { SwitcherComponent } from '../form/switcher/switcher.component';
import { TextareaDirective } from '../form/textarea/textarea.directive';
import { UploadComponent } from '../form/upload/upload.component';
import { GroupTitleComponent } from '../group-title/group-title.component';
import { GroupTitleModule } from '../group-title/group-title.module';
import {
    GuidedWorkflowComponent,
    GuidedWorkflowModule
} from '../guided-workflow';
import { IconComponent } from '../icon/icon.component';
import { IconModule } from '../icon/icon.module';
import { InfoButtonComponent, InfoButtonModule } from '../info-button';
import {
    LibraryComponent,
    LibraryGroupComponent,
    LibraryModule
} from '../library';
import {
    ListControllerActionComponent,
    ListControllerCategoryFilterComponent,
    ListControllerComponent,
    ListControllerFilterComponent,
    ListControllerGeneralComponent,
    ListControllerModule,
    ListControllerOrderComponent,
    ListControllerNumberOfResultsComponent
} from '../list-controller';
import {
    ListViewComponent,
    ListViewHeadlineComponent,
    ListViewLabelComponent,
    ListViewModule,
    ListViewSubheadlineComponent,
    ListViewValueComponent
} from '../list-view';
import { LoaderComponent, LoaderDirective, LoaderModule } from '../loader';
import { MainMenuComponent, MainMenuModule } from '../main-menu';
import { ModalModule } from '../modal';
import { DialogComponent } from '../modal/dialog/dialog.component';
import { DialogModule } from '../modal/dialog/dialog.module';
import { ProgressModalComponent } from '../modal/progress-modal/progress-modal.component';
import { ProgressModalModule } from '../modal/progress-modal/progress-modal.module';
import { NodeTreeComponent, NodeTreeModule } from '../node-tree/';
import { OverlayHelperService } from '../overlay/overlay-helper.service';
import { OverlayService } from '../overlay/overlay.service';
import { PaginationComponent, PaginationModule } from '../pagination';
import { PaneModule, TaskPaneComponent } from '../pane';
import { PanelItemComponent, PanelsComponent, PanelsModule } from '../panels';
import {
    RuleActionComponent,
    RuleEventComponent,
    RuleConditionComponent,
    RuleContainerComponent,
    RuleContainerModule,
    RuleTreeComponent
} from '../rule-container/';
import { SnackBarComponent, SnackBarModule } from '../snack-bar';
import { StateBadgeComponent, StateBadgeModule } from '../state-badge';
import { StickyModule } from '../sticky/sticky.module';
import { SystemBarComponent, SystemBarModule } from '../systembar';
import { TableComponent, TableRowComponent } from '../table';
import { TableRowModule } from '../table/table-row/table-row.module';
import { TableModule } from '../table/table.module';
import { TabsModule, TabsComponent, InkBarComponent } from '../tabs';
import { TactileBlockComponent } from '../tactile-block/tactile-block.component';
import { TactileBlockModule } from '../tactile-block/tactile-block.module';
import { TaskSearchComponent } from '../task-search/task-search.component';
import { TaskSearchModule } from '../task-search/task-search.module';
import {
    TileGroupComponent,
    TileItemComponent,
    TileSectionComponent,
    TileSectionMoreComponent
} from '../tile-group';
import { TileGroupModule } from '../tile-group/tile-group.module';
import { TooltipDirective, TooltipModule } from '../tooltip';
import { WizardComponent, WizardModule } from '../wizard';
import { ContentComponent } from '../wizard/step/content/content.component';
import { StepComponent } from '../wizard/step/step.component';
import { SummaryComponent } from '../wizard/step/summary/summary.component';
import { RainbowChartComponent } from '../chart/rainbow-chart/rainbow-chart.component';
import { RainbowChartModule } from '../chart/rainbow-chart/rainbow-chart.module';
import {
    ExternalTaskComponent,
    ExternalTaskModule
} from '../modal/external-task';
import { ToggleDirective, ToggleModule } from '../toggle';
import { TableSortingDirective } from '../table/table-sorting/table-sorting.directive';
import { TagComponent } from '../tags/tag/tag.component';
import { TagsComponent } from '../tags/tags.component';
import { TagsModule } from '../tags/tags.module';
import { TableCellMetaComponent } from '../table/table-row/table-cell/table-cell-meta/table-cell-meta.component';
import {
    DragAndDropModule,
    DroppableDirective,
    Draggable2Directive
} from '../drag-and-drop';
import { StackedBarChartModule } from '../chart/stacked-bar-chart/stacked-bar-chart.module';
import { StackedBarChartComponent } from '../chart/stacked-bar-chart/stacked-bar-chart.component';
import { ThemeService } from '../theme/theme.service';
import {
    BucketVisualisationModule,
    BucketVisualisationComponent
} from '../bucket-visualisation';
import {
    UniversalModule,
    UniversalAppsComponent,
    UniversalTabsComponent,
    UniversalOverlayComponent,
    UniversalService
} from '../universal';
import { MenuCardComponent, MenuCardModule } from '../menu-card';
import { BatchActionsComponent, BatchActionsModule } from '../batch-actions';
import { TimelineComponent, TimelineModule } from '../timeline';
import { PageService } from '../../core/services/page.service';
import { AnnotationModule, AnnotationComponent } from '../annotation';
import { BlockDropElementDirective } from '../block-tree/block-dropelement.directive';
import { CharacteristicsComponent, CharacteristicsModule } from '../characteristics';

import { SelectCellComponent } from '../table2/table2-row/select-cell/select-cell.component';
import { SelectHeaderComponent } from '../table2/header/select-header/select-header.component';
import { TableColumnSetSelectorMenuComponent } from '../table2/head/column-set-selector-menu/column-set-selector-menu.component';
import { SingleActionsCellComponent } from '../table2/table2-row/single-actions-cell/single-actions-cell.component';
import { TableHeadComponent } from '../table2/head/head.component';
import { TableTitleComponent } from '../table2/head/head-title/title.component';
import { TableTitleCountComponent } from '../table2/head/head-title/head-count/count.component';
import { TableHeadListControlsComponent } from '../table2/head/head-list-controls/list-controls.component';
import { TableHeadActionsComponent } from '../table2/head/head-actions/actions.component';
import { TableBatchActionsComponent } from '../table2/head/head-batch-actions/batch-actions.component';
import { DraggableService } from '../draggable/draggable.service';
import { TableHeaderLabelComponent } from '../table2/header/header-label/header-label.component';
import { PipesModule, CuiHighlightPipe } from '../../core/pipes';
import { Table2Component } from '../table2/table.component';
import { HeaderComponent } from '../table2/header/header.component';
import { Row2Component } from '../table2/table2-row/row.component';
import { RowTemplateDirective } from '../table2/row-template/row-template.directive';
import { CellDirective } from '../table2/table2-row/cell/cell.directive';
import { Table2Module } from '../table2/table.module';
import { TableRowChildCountComponent } from '../table2/table2-row/child-count/child-count.component';
import { TablePaginationComponent } from '../table2/table-pagination/table-pagination.component';
import { TogglerComponent } from './../toggler/toggler.component';
import { TogglerModule } from './../toggler/toggler.module';
import { TogglerBreakPointDirective } from '../toggler/toggler-breakpoint.directive';
import { TogglerCollapseComponent } from '../toggler/toggler-collapse/toggler-collapse.component';
import {
    ListControllerClearFiltersComponent
} from '../list-controller/list-controller-clear-filters/list-controller-clear-filters.component';

import { ValidationSummaryService } from '../validation-summary/validation-summary.service';
import { ValidationSummaryModule } from '../validation-summary/validation-summary.module';
import { ValidationSummaryComponent } from '../validation-summary/validation-summary.component';
import {
    RelationGraphModule,
    RelationGraphComponent
} from '../relation-graph';
import { TableSingleSelectCellComponent } from '../table2/table2-row/single-select-cell/single-select-cell.component';
import { TableCellIndexComponent } from '../table2/table2-row/index/index.component';
import { ExportModule } from '../export/export.module';
import { OdenComboboxComponent } from '../oden-form/oden-combobox/oden-combobox.component';
import { OdenErrorComponent } from '../oden-form/oden-error/oden-error.component';
import { OdenFormFieldComponent } from '../oden-form/oden-form-field/oden-form-field.component';
import { OdenFormFieldsetComponent } from '../oden-form/oden-form-fieldset/oden-form-fieldset.component';
import { OdenFormGroupComponent } from '../oden-form/oden-form-group/oden-form-group.component';
import { OdenLabelDirective } from '../oden-form/oden-label/oden-label.directive';
import { OdenFormHeaderDirective } from '../oden-form/oden-form-header/oden-form-header.directive';
import { OdenInputDirective } from '../oden-form/oden-input/oden-input.directive';
import { OdenOptionComponent } from '../oden-form/oden-combobox/oden-option/oden-option.component';
import { OdenRadioDirective } from '../oden-form/oden-radio/oden-radio.directive';
import { OdenSwitchDirective } from '../oden-form/oden-switch/oden-switch.directive';
import { OdenTextareaDirective } from '../oden-form/oden-textarea/oden-textarea.directive';
import { OdenWarningDirective } from '../oden-form/oden-warning/oden-warning.directive';
import { PreventDraggingModule } from '../prevent-dragging/prevent-dragging.module';
import { IndicateDraggingModule } from '../indicate-dragging/indicate-dragging.module';
import { LibraryItemReceiverService } from '../library/library-item-receiver.service';
import { ImageContainerComponent } from '../image-container/image-container.component';
import { ImageAttributionDirective } from '../image-container/attribution/attribution.directive';
import { ImageContainerModule } from '../image-container/image-container.module';

// Components that should be exposed to rm/application level
const MODULE_COMPONENTS_EXPORTS = [
    AutoScrollDirective,
    ActionMenuComponent,
    ActionMenuItemComponent,
    AnnotationComponent,
    AppHeaderComponent,
    BarChartComponent,
    BatchActionsComponent,
    BlockComponent,
    BlockDropElementDirective,
    BlockTreeComponent,
    BlockTreeBranchComponent,
    BlockTreeBranchLeafComponent,
    ButtonComponent,
    ButtonWrapperDirective,
    BucketVisualisationComponent,
    CallOutComponent,
    CheckboxDirective,
    ColumnComponent,
    ColumnSizeDirective,
    ContainerComponent,
    ContentComponent,
    CuiHighlightPipe,
    DateInputComponent,
    DatePickerComponent,
    DateTimeInputComponent,
    DateTimePickerComponent,
    DateTimeValidatorDirective,
    DateValidatorDirective,
    DraggableDirective,
    Draggable2Directive,
    DrawerComponent,
    MultiDrawerComponent,
    DroppableDirective,
    DropzoneComponent,
    DynamicSelectComponent,
    ElementOverlayDirective,
    EntityVersionComponent,
    ErrorPageComponent,
    ExternalTaskComponent,
    GaugeChartComponent,
    GroupTitleComponent,
    GuidedWorkflowComponent,
    IconComponent,
    InfoButtonComponent,
    InputDirective,
    InputSearchDirective,
    LibraryComponent,
    LibraryGroupComponent,
    LineChartComponent,
    ListControllerActionComponent,
    ListControllerCategoryFilterComponent,
    ListControllerComponent,
    ListControllerFilterComponent,
    ListControllerGeneralComponent,
    ListControllerNumberOfResultsComponent,
    ListControllerClearFiltersComponent,
    ListControllerOrderComponent,
    ListViewComponent,
    ListViewHeadlineComponent,
    ListViewLabelComponent,
    ListViewSubheadlineComponent,
    ListViewValueComponent,
    LoaderComponent,
    LoaderDirective,
    MainMenuComponent,
    MenuCardComponent,
    MultiButtonComponent,
    NodeTreeComponent,
    OdenCheckboxDirective,
    OdenComboboxComponent,
    OdenErrorComponent,
    OdenFormFieldComponent,
    OdenFormFieldsetComponent,
    OdenFormGroupComponent,
    OdenFormHeaderDirective,
    OdenInputDirective,
    OdenLabelDirective,
    OdenOptionComponent,
    OdenRadioDirective,
    OdenSwitchDirective,
    OdenTextareaDirective,
    OdenWarningDirective,
    OptionComponent,
    PaginationComponent,
    PanelItemComponent,
    PanelsComponent,
    PeriodPickerComponent,
    PieChartComponent,
    PipesModule,
    ProgressBarComponent,
    ProgressModalComponent,
    RadioDirective,
    RainbowChartComponent,
    RequiredDirective,
    RowComponent,
    RuleActionComponent,
    RuleConditionComponent,
    RuleContainerComponent,
    RuleTreeComponent,
    RuleEventComponent,
    SelectComponent,
    SimpleComponent,
    SliderComponent,
    StateBadgeComponent,
    StackedBarChartComponent,
    StepComponent,
    SummaryComponent,
    SwitcherComponent,
    SystemBarComponent,
    TableComponent,
    TableRowComponent,
    TableCellMetaComponent,
    TactileBlockComponent,
    TagComponent,
    TagsComponent,
    TaskPaneComponent,
    TaskSearchComponent,
    TextareaDirective,
    TileGroupComponent,
    TileItemComponent,
    TileSectionComponent,
    TileSectionMoreComponent,
    TimeInputComponent,
    TimelineComponent,
    TimePickerComponent,
    TimeValidatorDirective,
    TooltipDirective,
    ToggleDirective,
    UploadComponent,
    UniversalTabsComponent,
    UniversalAppsComponent,
    UniversalOverlayComponent,
    WizardComponent,
    TableSortingDirective,
    TabsComponent,
    InkBarComponent,
    CharacteristicsComponent,
    Table2Component,
    HeaderComponent,
    Row2Component,
    RowTemplateDirective,
    CellDirective,
    SelectCellComponent,
    SelectHeaderComponent,
    TableColumnSetSelectorMenuComponent,
    SingleActionsCellComponent,
    TableHeadComponent,
    TableTitleComponent,
    TableTitleCountComponent,
    TableRowChildCountComponent,
    TableHeadListControlsComponent,
    TableHeadActionsComponent,
    TableBatchActionsComponent,
    TableHeaderLabelComponent,
    TablePaginationComponent,
    TogglerComponent,
    TogglerBreakPointDirective,
    TogglerExpandComponent,
    TogglerCollapseComponent,
    ValidationSummaryComponent,
    TableSingleSelectCellComponent,
    TableCellIndexComponent,
    RelationGraphComponent,
    ImageContainerComponent,
    ImageAttributionDirective
];

// Imported modules
const MODULE_IMPORTS = [
    ActionMenuModule.forRoot(),
    AddonModule.forRoot(),
    AnnotationModule.forRoot(),
    AppHeaderModule.forRoot(),
    BackdropModule.forRoot(),
    BarChartModule.forRoot(),
    BatchActionsModule.forRoot(),
    StackedBarChartModule.forRoot(),
    BlockModule.forRoot(),
    BlockTreeModule.forRoot(),
    BreadcrumbModule.forRoot(),
    ButtonModule.forRoot(),
    BucketVisualisationModule.forRoot(),
    CallOutModule.forRoot(),
    ContainerModule.forRoot(),
    DialogModule.forRoot(),
    DragAndDropModule.forRoot(),
    DraggableModule.forRoot(),
    DrawerModule.forRoot(),
    DropzoneModule.forRoot(),
    ElementOverlayModule.forRoot(),
    EntityVersionModule.forRoot(),
    ErrorPageModule.forRoot(),
    ExternalTaskModule.forRoot(),
    FormModule.forRoot(),
    GaugeChartModule.forRoot(),
    GroupTitleModule.forRoot(),
    GuidedWorkflowModule.forRoot(),
    IconModule.forRoot(),
    InfoButtonModule.forRoot(),
    LibraryModule.forRoot(),
    LineChartModule.forRoot(),
    ListControllerModule.forRoot(),
    ListViewModule.forRoot(),
    LoaderModule.forRoot(),
    MainMenuModule.forRoot(),
    MenuCardModule.forRoot(),
    ModalModule.forRoot(),
    NodeTreeModule.forRoot(),
    OdenFormModule.forRoot(),
    PaginationModule.forRoot(),
    PaneModule.forRoot(),
    PanelsModule.forRoot(),
    PieModule.forRoot(),
    PipesModule.forRoot(),
    ProgressModalModule.forRoot(),
    RainbowChartModule.forRoot(),
    RuleContainerModule.forRoot(),
    SimpleModule.forRoot(),
    SnackBarModule.forRoot(),
    StateBadgeModule.forRoot(),
    StickyModule.forRoot(),
    SystemBarModule.forRoot(),
    TableModule.forRoot(),
    TableRowModule.forRoot(),
    TabsModule.forRoot(),
    TactileBlockModule.forRoot(),
    TagsModule.forRoot(),
    TaskSearchModule.forRoot(),
    TileGroupModule.forRoot(),
    TimelineModule.forRoot(),
    TooltipModule.forRoot(),
    ToggleModule.forRoot(),
    TogglerModule.forRoot(),
    UniversalModule.forRoot(),
    WizardModule.forRoot(),
    CharacteristicsModule.forRoot(),
    Table2Module.forRoot(),
    ValidationSummaryModule.forRoot(),
    RelationGraphModule.forRoot(),
    ExportModule.forRoot(),
    PreventDraggingModule.forRoot(),
    IndicateDraggingModule.forRoot(),
    ImageContainerModule.forRoot()
];

@NgModule({
    imports: MODULE_IMPORTS,
    exports: MODULE_COMPONENTS_EXPORTS,
    providers: [
        FilterService,
        OverlayHelperService,
        PageService,
        VersionsService,
        { provide: 'windowObject', useValue: window },
        ResizeService
    ],
    entryComponents: [
        BackdropComponent,
        DateTimePickerComponent,
        DialogComponent,
        LoaderComponent,
        ProgressModalComponent,
        SnackBarComponent
    ]
})
export class OdenModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: OdenModule,
            providers: [
                ApplicationService,
                ThemeService,
                OverlayService,
                PostMessageService,
                UniversalService,
                AppHeaderService,
                DraggableService,
                ValidationSummaryService,
                WhitelabelService,
                LibraryItemReceiverService
            ]
        };
    }
}
