// @ericsson/oden/components/tile-group/tile-group.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';

import { ActionMenuModule } from '../action-menu/action-menu.module';
import { StateBadgeModule } from '../state-badge/state-badge.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { TileGroupComponent } from './tile-group.component';
import { TileItemComponent } from './tile-item/tile-item.component';
import { TileSectionComponent } from './tile-item/tile-section/tile-section.component';
import { TileSectionMoreComponent } from './tile-item/tile-section/tile-section-more/tile-section-more.component';
import { IconModule } from '../icon/icon.module';
import { TableModule } from '../table/table.module';
import { FormModule } from '../form/form.module';
import { BatchActionsModule } from '../batch-actions/batch-actions.module';
import { TimelineModule } from '../timeline/timeline.module';
import { ElementOverlayModule } from '../element-overlay/element-overlay.module';
import { TileGroupService } from './tile-group.service';
import { TagsModule } from '../tags/tags.module';
import { IndicateDraggingModule } from '../indicate-dragging/indicate-dragging.module';

@NgModule({
    imports: [CommonModule, ActionMenuModule, TooltipModule, StateBadgeModule, IconModule, TranslateModule.forChild(),
        TableModule, FormModule, BatchActionsModule, TimelineModule, ElementOverlayModule, TagsModule, IndicateDraggingModule],
    exports: [TileGroupComponent, TileItemComponent, TileSectionComponent, TileSectionMoreComponent],
    declarations: [TileGroupComponent, TileItemComponent, TileSectionComponent, TileSectionMoreComponent]
})

export class TileGroupModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: TileGroupModule,
            providers: [TileGroupService]
        };
    }
}
