// @ericsson/cus-ui/task/task.module.ts
import { CommonModule } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, Injector, ModuleWithProviders, NgModule } from '@angular/core';
import { OdenModule } from '@ericsson/oden/modules';

import { GUIDE_ROUTE_PATH } from './guide/services/guide.service';
import { TaskModuleOptions } from './interface/task-module-options';
import { OrchestratorService } from './orchestrator.service';
import { TaskScreenService } from './task.screen.service';
import { UserJourneyService } from './user-journey.service';

@NgModule({
    imports: [CommonModule, OdenModule],
    providers: [],
    entryComponents: [],
    schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class TaskModule {

    static injector: Injector;
    static forRoot(options?: TaskModuleOptions): ModuleWithProviders {
        return {
            ngModule: TaskModule,
            providers: [
                OrchestratorService,
                TaskScreenService,
                UserJourneyService,
                {
                    provide: GUIDE_ROUTE_PATH,
                    useValue: options && options.routePath ? options.routePath : 'guidedtask'
                }
            ]
        }
    }

    constructor(public injector: Injector) {
        TaskModule.injector = injector;
    }
}
