// @ericsson/cus-ui/task/task.constants.ts
export class TaskConstants {
    static GUIDE_ROUTE_PATH: string = 'guidedtask';
    static SUMMARY_TASK_ID: string = 'summary';
    static SUMMARY_TASK_SCREEN_PATH: string = 'summary';
    static FINISH_TITLE: string = 'Finish';
    static PAUSE_TASK_ACTION: string = 'pause';
    static DISCARD_TASK_ACTION: string = 'discard';
    static EDIT_TASK_ACTION: string = 'edit';
    static PAUSE_TASK_MESSAGE: string = `The task will be paused. Do you want to save the changes
    before pausing this task?<br/><br/>
    If you don't save, your changes made to this task will be lost.`;
    static EDIT_TASK_MESSAGE: string = `Do you want to edit this task?`;
    static PUBLISH_TASK_MESSAGE: string = `Do you want to finish this task?`;
    static TASK_NOT_STARTED: string = 'NotStarted';
    static PAUSE_TASK_LABEL: string = 'Pause Task';
    static DISCARD_TASK_LABEL: string = 'Discard Task';
    static EDIT_TASK_LABEL: string = 'Edit Task';
    static NEXT_TASK_ACTION: string = 'Next';
    static CANCEL_LABEL: string = 'Cancel';
    static CANCEL_ACTION: string = 'cancel';
    static DONT_SAVE_LABEL: string = 'Don\'t Save';
    static DONT_SAVE_ACTION: string = 'dont save';
    static SAVE_LABEL: string = 'Save';
    static SAVE_ACTION: string = 'save';
    static OK_LABEL: string = 'OK';
    // note: Don't keep simple 'ok' alone, add some prefix to avoid navigating away from guide view during application error.
    // since httpInterceptor has same action name of 'ok'
    static OK_ACTION: string = 'task_ok';
    static OK_ERROR_ACTION: string = 'error_ok';
    static PUBLISH_OK_ERROR_ACTION: string = 'publish_error_ok';
    static EDIT_OK_ACTION: string = 'edit_ok';
    static PUBLISH_OK_ACTION: string = 'publish_ok';
    static EXIT_ALERT: string = 'Do you want to exit from this task? On confirmation, all your saved/unsaved changes will be lost.';
    static NO_LABEL: string = 'No';
    static NO_ACTION: string = 'no';
    static YES_LABEL: string = 'Yes';
    static YES_ACTION: string = 'yes';
    static DONT_DISCARD_LABEL: string = 'Don\'t Discard';
    static ASSIGNED_TO: string = 'Assigned to';
    static OWNER: string = 'Owner';
    static DUE_DATE: string = 'Due Date';
    static TASKS_ERROR_LABEL: string = 'Server Error';
    static TASKS_ERROR_TEXT_1: string = 'Server is temporarily unavailable, '
    static TASKS_ERROR_TEXT_2: string = `please retry. If problem continues, please contact your system administrator`;
    static RETRY_LABEL: string = 'Retry';
    static ASSIGNED_ACTION: string = 'assigned';
    static NEW_ACTION: string = 'new';
    static PAUSED_ACTION: string = 'paused';
    static MY_TASKS_LABEL: string = 'My Tasks';
    static ALL_TASKS_LABEL: string = 'All Tasks';
    static CUS_REGISTRATION: string = 'CUS';
    static TASK_REGISTRATION_ERROR: string = 'TaskListItem Registration not found.';
    static REQUIRED_FIELD_MISSING: string = 'Required field is missing from the response.';
    static DISCARD_MESSAGE_PART1: string = 'Do you want to discard';
    static DISCARD_MESSAGE_PART2: string = 'task?';
    static DISCARDED_MESSAGE: string = 'Your task has been discarded and you cannot proceed with this task';
    static DISCARD_SNACKBAR_SUCCESS_MESSAGE: string = 'has been deleted successfully!';
    static DISCARD_SNACKBAR_FAILURE_MESSAGE: string = 'could not be deleted!';
    static NO_RECORDS_MESSAGE: string = 'No records available';
    static MYTASKS_NO_RECORDS_SUB_MESSAGE: string = 'Please go to all tasks for assigning a task';
    static ALLTASKS_NO_RECORDS_SUB_MESSAGE: string = 'Please contact your system administrator';
    static NO_MATCHING_RESULTS_MESSAGE: string = 'No matching results found';
    static NO_COLUMNS_SELECTED_MESSAGE: string = 'No columns are selected';
    static TASK_LABEL: string = 'Task';
    static WORKFLOW_LABEL: string = 'Workflow';
    static STATE_LABEL: string = 'State';
    static RESUME_LABEL: string = 'Resume';
    static DETAILS_LABEL: string = 'Details';
    static DISCARD_LABEL: string = 'Discard';
    static ALL_LABEL: string = 'All';
    static STEP_LABEL: string = 'Step';
    static USER_TASK_LABEL: string = 'User Task';
    static DESCRIPTION_LABEL: string = 'Description';
    static GUIDE_POSITION: string = 'bottom';
    static PUBLISH_WORKFLOW: string = 'Task';
    static PUBLISH_WORKFLOW_FINISH: string = 'is finished';
    static PUBLISH_WORKFLOW_COMPLETE: string = 'All the tasks are succesfully finished';
    static DATE_FORMAT: string = 'd MMM y hh:mm:ss:SS';
}
