// @ericsson/oden/components/wizard/wizard.module.ts
import { NgModule, ModuleWithProviders } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';
import { CommonModule } from '@angular/common';
import { WizardComponent } from './wizard.component';
import { SummaryComponent } from './step/summary/summary.component';
import { ContentComponent } from './step/content/content.component';
import { StepComponent } from './step/step.component';
import { ButtonModule } from '../button/button.module';
import { IconModule } from '../icon/icon.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { StepContainerComponent } from './step/step-container/step-container.component';

@NgModule({
    imports: [CommonModule, ButtonModule, IconModule, TooltipModule, TranslateModule.forChild()],
    exports: [WizardComponent, SummaryComponent, ContentComponent, StepComponent, StepContainerComponent],
    declarations: [WizardComponent, SummaryComponent, ContentComponent, StepComponent, StepContainerComponent]
})

export class WizardModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: WizardModule
        };
    }
}
