// @ericsson/oden/components/tactile-block/tactile-block.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';

import { ActionMenuModule } from '../action-menu/action-menu.module';
import { ButtonModule } from '../button/button.module';
import { DraggableModule } from '../draggable/draggable.module';
import { IconModule } from '../icon/icon.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { TactileBlockComponent } from './tactile-block.component';
import { TactileBlockService } from './tactile-block.service';
import { FormModule } from '../form';
import { PreventDraggingModule } from '../prevent-dragging/prevent-dragging.module';
import { TranslateModule } from '@ngx-translate/core';
import { AnnotationModule } from '../annotation';
import { IndicateDraggingModule } from '../indicate-dragging/indicate-dragging.module';

@NgModule({
    imports: [CommonModule, ActionMenuModule, IconModule, ButtonModule, PreventDraggingModule, IndicateDraggingModule,
        DraggableModule, TooltipModule, FormModule, AnnotationModule, TranslateModule.forChild()],
    exports: [TactileBlockComponent],
    declarations: [TactileBlockComponent]
})

export class TactileBlockModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: TactileBlockModule,
            providers: [TactileBlockService],
        };
    }
}
