// @ericsson/oden/components/chart/addon/addon.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { ContainerModule } from '../../container';
import { LegendAddonModule } from './legend/legend-addon.module';
import { SumAddonModule } from './sum/sum-addon.module';
import { LinkAddonComponent } from './link/link-addon.component';
import { LegendAddonComponent } from './legend/legend-addon.component';
import { TrendAddonComponent } from './trend/trend-addon.component';
import { TrendAddonModule } from './trend/trend-addon.module';
import { LinkAddonModule } from './link/link-addon.module';
import { SumAddonComponent } from './sum/sum-addon.component';
import { SelectorAddonComponent, SelectorAddonModule } from './selector';

@NgModule({
    imports: [CommonModule, ContainerModule,
        LinkAddonModule, LegendAddonModule, SumAddonModule, TrendAddonModule, SelectorAddonModule],
    exports: [LinkAddonComponent, LegendAddonComponent, TrendAddonComponent, SumAddonComponent, SelectorAddonComponent],
    entryComponents: [LinkAddonComponent, LegendAddonComponent, TrendAddonComponent, SumAddonComponent, SelectorAddonComponent]
})

export class AddonModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: AddonModule
        };
    }
}
