// @ericsson/oden/core/constants/bar-chart-config.constants.ts
export const BARCHARTCONFIG: BarChartConfig = {
    /**
     *  Cofiguration for bar chart.
     *  MARGINS: sets the inner margins of the SVG element
     *  MAXBARS: the maximum amount of models to display as bars
     *  MAXSTACKEDBARS: bar must support displaying up to 8 sub-categories.
     *  DURATION: transition speed
     *  MAX_VERTICAL_BARWIDTH: Maximum width for each vertical bar width
     *  MAX_HORIZONTAL_BARWIDTH: Maximum width for each horizontal bar width
     */
    MARGINS: { top: 30, right: 30, bottom: 20, left: 40 },
    MAXBARS: 12,
    MAX_HORIZONTAL_BARS: 5,
    DURATION: 500,
    MAXSTACKEDBARS: 7,
    BARRADIUS: 2,
    OPACITY: .65,
    AXIS_TITLE_OPACITY: .45,
    DEFAULT_TICKS: 5,
    MAX_VERTICAL_BARWIDTH: 48,
    MAX_HORIZONTAL_BARWIDTH: 48,
    MIN_HORIZONTAL_BARWIDTH: 12,
    MIN_VERTICAL_BARWIDTH: 12
};

export type BarChartMargins = {
    top: number,
    right: number,
    bottom: number,
    left: number,
};

export type BarChartConfig = {
    MARGINS: BarChartMargins,
    MAXBARS: number,
    MAX_HORIZONTAL_BARS: number,
    DURATION: number,
    MAXSTACKEDBARS: number,
    BARRADIUS: number,
    OPACITY: number,
    AXIS_TITLE_OPACITY: number,
    DEFAULT_TICKS: number,
    MAX_VERTICAL_BARWIDTH: number,
    MAX_HORIZONTAL_BARWIDTH: number,
    MIN_VERTICAL_BARWIDTH: number,
    MIN_HORIZONTAL_BARWIDTH: number
};
