// @ericsson/oden/core/constants/color.constants.ts
import { ColorHelper } from '../helpers/color.helper';

export class COLOR {
    // Brand 2.0
    // Ericsson Main Brand Colors
    static ERICSSON_DODGERBLUE: string = '#0082F0';
    static ERICSSON_LOCHMARA: string = '#2B77AA';
    static ERICSSON_AMETHYST: string = '#A56EBE';
    static ERICSSON_ELM: string = '#168277';

    // Ericsson Tinted Main Brand Colors
    static ERICSSON_TINTED_DODGERBLUE: string = '#2995F0';
    static ERICSSON_TINTED_LOCHMARA: string = '#2974A8';
    static ERICSSON_TINTED_AMETHYST: string = '#B284C7';
    static ERICSSON_TINTED_ELM: string = '#3B958C';

    // Ericsson Status Colours
    static ERICSSON_OLDROSE: string = '#C7324F';
    static ERICSSON_REDORANGE: string = '#F52323';
    static ERICSSON_CARROTORANGE: string = '#F78521';
    static ERICSSON_MOONYELLOW: string = '#F2C618';
    static ERICSSON_JUNGLEGREEN: string = '#36B07F';

    // Ericsson tinted Status Colours
    static ERICSSON_LIGHTER_OLDROSE: string = '#DC3967';
    static ERICSSON_LIGHTER_REDORANGE: string = '#F44646';
    static ERICSSON_LIGHTER_CARROTORANGE: string = '#F69744';
    static ERICSSON_LIGHTER_MOONYELLOW: string = '#F2CD3D';
    static ERICSSON_LIGHTER_JUNGLEGREEN: string = '#56BB92';

    static BLUE: string = '#007BE4';
    static GREEN: string = '#56BB92';
    static BLACK: string = '#000000';
    static WHITE: string = '#FFFFFF';
    static LIGHTGRAY4: string = '#CBCCCD';
    static LIGHTGRAY3: string = '#D8D9DA';
    static LIGHTGRAY2: string = '#BEC0C1';

    // Ericsson Gray
    static ERICSSON_BLACK: string = '#181818';
    static ERICSSON_GRAY1: string = '#242424';
    static ERICSSON_GRAY2: string = '#767676';
    static ERICSSON_GRAY3: string = '#a0a0a0';
    static ERICSSON_GRAY4: string = '#e0e0e0';
    static ERICSSON_GRAY5: string = '#f2f2f2';
    static ERICSSON_WHITE: string = '#fafafa';

    // Other colors
    /* The Chart Colour Loop*/
    static BRAND_COLOR_1: string = COLOR.ERICSSON_DODGERBLUE;
    static BRAND_COLOR_2: string = COLOR.ERICSSON_LOCHMARA;
    static BRAND_COLOR_3: string = COLOR.ERICSSON_AMETHYST;
    static BRAND_COLOR_4: string = COLOR.ERICSSON_ELM;

    static DARK_YELLOW: string = COLOR.ERICSSON_LIGHTER_MOONYELLOW;
    static ORANGE: string = COLOR.ERICSSON_LIGHTER_CARROTORANGE;
    static RED: string = COLOR.ERICSSON_LIGHTER_REDORANGE;
    static DARK_RED: string = COLOR.ERICSSON_OLDROSE;

    static LIGHT_GREEN: string = '#CDDFDD';
    static LIGHT_BLUE: string = '#E6ECF2';

    static DARK_GRAY: string = '#2B2B2B';
    static GRAY: string = '#BEC0C1';

    // Status colors
    static BRAND_SUCCESS: string = COLOR.ERICSSON_LIGHTER_JUNGLEGREEN;
    static BRAND_NOTICE: string = COLOR.ERICSSON_LIGHTER_MOONYELLOW;
    static BRAND_WARNING: string = COLOR.ERICSSON_LIGHTER_CARROTORANGE;
    static BRAND_ERROR: string = COLOR.ERICSSON_LIGHTER_REDORANGE;
    static BRAND_CRITICAL: string = COLOR.ERICSSON_LIGHTER_OLDROSE;
    static BRAND_PUBLISH: string = COLOR.BLUE;
    static BRAND_DRAFT: string = COLOR.LIGHT_BLUE;

    static SUCCESS: string = COLOR.BRAND_SUCCESS;
    static NOTICE: string = COLOR.BRAND_NOTICE;
    static WARNING: string = COLOR.BRAND_WARNING;
    static ERROR: string = COLOR.BRAND_ERROR;
    static CRITICAL: string = COLOR.BRAND_CRITICAL;
    static PUBLISH: string = COLOR.BRAND_PUBLISH;
    static DRAFT: string = COLOR.BRAND_DRAFT;

    // DARK_THEME
    static DARKTHEME_DARKEST: string = '#161616';
    static DARKTHEME_DARKER: string = '#212121';
    static DARKTHEME_DARK: string = '#303030';

    /* Lifecycles Color */
    static INITIATED: string = '#ABD6FA';
    static GETTING_THERE: string = '#80C1Fa';
    static RUNNING: string = COLOR.ERICSSON_DODGERBLUE;
    static STOPPED: string = COLOR.LIGHTGRAY3;
    static END: string = COLOR.LIGHTGRAY3;

    static BRAND_COLORS = [
        COLOR.BRAND_COLOR_1,
        COLOR.BRAND_COLOR_2,
        COLOR.BRAND_COLOR_3,
        COLOR.BRAND_COLOR_4,
    ];

    static BRAND_STATUSCOLORS = [
        COLOR.SUCCESS,
        COLOR.NOTICE,
        COLOR.WARNING,
        COLOR.ERROR,
        COLOR.CRITICAL,
    ];

    static BRAND_SEVERITYCOLORS = [
        COLOR.LIGHT_GREEN,
        COLOR.DARK_YELLOW,
        COLOR.ORANGE,
        COLOR.RED,
        COLOR.DARK_RED,
    ];

    /** @deprecated */
    static tintColor(color: string, percent: number): string {
        return ColorHelper.tintColor(color, percent);
    }

    /** @deprecated */
    static getBrandColors = () => COLOR.BRAND_COLORS;

    /** @deprecated */
    static graphBrandColorByIndex = (index: number): [TintedColor, BaseColor] => {
        return ColorHelper.graphBrandColorByIndex(index);
    }

    /** @deprecated */
    static graphColorByIndex = (index: number, color: Array<string>): [TintedColor, BaseColor] => {
        return ColorHelper.graphColorByIndex(index, color);
    }
    static BRAND_PRIMARY = 'brand-primary';
    static BRAND_SECONDARY = 'brand-secondary';
    static BRAND_THIRD = 'brand-third';
    static BRAND_FORTH = 'brand-forth';
}

type TintedColor = string;
type BaseColor = string;
