// @ericsson/oden/core/constants/color.whitelabel.ts

export class COLORWHITELABEL {
    static SUCCESS: string = 'status-success';
    static NOTICE: string = 'status-notice';
    static WARNING: string = 'status-warning';
    static ERROR: string = 'status-error';
    static CRITICAL: string = 'status-critical';
    static PUBLISH: string = 'brand-publish';
    static DRAFT: string = 'brand-draft';
    static MINOR: string = 'status-minor';
    static BRAND_PRIMARY = 'brand-primary';
    static BRAND_SECONDARY = 'brand-secondary';
    static BRAND_THIRD = 'brand-third';
    static BRAND_FORTH = 'brand-forth';

    static BASE_COLOR = 'base-color';
    static BASE_COLOR_INVERTED = 'base-color-inverted';

    // Opaque levels
    static OPAQUE_100 = 1;
    static OPAQUE_75 = 0.75;
    static OPAQUE_60 = 0.60;
    static OPAQUE_40 = 0.40;
    static OPAQUE_20 = 0.20;
    static OPAQUE_15 = 0.15;
    static OPAQUE_5 = 0.05;
    static OPAQUE_0 = 0;

    // Tints levels
    static TINT_20 = 0.2;
    static TINT_17 = 0.17;
    static TINT_10 = 0.1;
    static TINT_5 = 0.05;

    static GRAY = 'gray';
    static GRAY2 = 'gray2';
    static GRAY3 = 'gray3';
    static GRAY4 = 'gray4';
    static GRAY5 = 'gray5';
    static GRAY6 = 'gray6';
    static GRAY7 = 'gray7';
    static GRAY8 = 'gray8';

    static C333333 = 'C333333';

}
