// @ericsson/oden/components/table2/table.module.ts
import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Table2Component } from './table.component';
import { FormsModule } from '@angular/forms';
import { HeaderComponent } from './header/header.component';
import { Row2Component } from './table2-row/row.component';
import { CellDirective } from './table2-row/cell/cell.directive';
import { RowTemplateDirective } from './row-template/row-template.directive';
import { FormModule } from '../form/form.module';
import { SelectCellComponent } from './table2-row/select-cell/select-cell.component';
import { SelectHeaderComponent } from './header/select-header/select-header.component';
import { IconModule } from '../icon/icon.module';
import { SingleActionsCellComponent } from './table2-row/single-actions-cell/single-actions-cell.component';
import { ActionMenuModule } from '../action-menu/action-menu.module';
import { TableHeadComponent } from './head/head.component';
import { TableTitleComponent } from './head/head-title/title.component';
import { TableTitleCountComponent } from './head/head-title/head-count/count.component';
import { TableHeadListControlsComponent } from './head/head-list-controls/list-controls.component';
import { TableHeadActionsComponent } from './head/head-actions/actions.component';
import { TableBatchActionsComponent } from './head/head-batch-actions/batch-actions.component';
import { BatchActionsModule } from '../batch-actions/batch-actions.module';
import { TranslateModule } from '@ngx-translate/core';
import { TableColumnSetSelectorComponent } from './head/column-set-selector/column-set-selector.component';
import { TableHeaderLabelComponent } from './header/header-label/header-label.component';
import { TooltipModule } from '../tooltip/tooltip.module';
import { ContainerModule } from '../container/container.module';
import { TableRowChildCountComponent } from './table2-row/child-count/child-count.component';
import { TablePaginationComponent } from './table-pagination/table-pagination.component';
import { PaginationModule } from '../pagination';
import { TableColumnSetSelectorMenuComponent } from './head/column-set-selector-menu/column-set-selector-menu.component';
import { TableSingleSelectCellComponent } from './table2-row/single-select-cell/single-select-cell.component';
import { TableCellIndexComponent } from './table2-row/index/index.component';
import { ModalModule } from '../modal';
import { ExportModule } from '../export/export.module';

@NgModule({
    imports: [
        CommonModule,
        FormModule,
        FormsModule,
        IconModule,
        ActionMenuModule,
        BatchActionsModule,
        TranslateModule,
        TooltipModule,
        ContainerModule,
        PaginationModule,
        BatchActionsModule,
        TranslateModule,
        TooltipModule,
        ModalModule,
        ExportModule
    ],
    declarations: [
        Table2Component,
        HeaderComponent,
        Row2Component,
        RowTemplateDirective,
        CellDirective,
        SelectCellComponent,
        SelectHeaderComponent,
        SingleActionsCellComponent,
        TableHeadComponent,
        TableTitleComponent,
        TableTitleCountComponent,
        TableRowChildCountComponent,
        TableHeadListControlsComponent,
        TableHeadActionsComponent,
        TableBatchActionsComponent,
        TableColumnSetSelectorComponent,
        TableColumnSetSelectorMenuComponent,
        TableHeaderLabelComponent,
        TablePaginationComponent,
        TableSingleSelectCellComponent,
        TableCellIndexComponent
    ],
    exports: [
        Table2Component,
        HeaderComponent,
        Row2Component,
        RowTemplateDirective,
        CellDirective,
        SelectCellComponent,
        SelectHeaderComponent,
        SingleActionsCellComponent,
        TableHeadComponent,
        TableTitleComponent,
        TableTitleCountComponent,
        TableRowChildCountComponent,
        TableHeadListControlsComponent,
        TableHeadActionsComponent,
        TableBatchActionsComponent,
        TableHeaderLabelComponent,
        TablePaginationComponent,
        TableColumnSetSelectorMenuComponent,
        TableSingleSelectCellComponent,
        TableCellIndexComponent
    ]
})
export class Table2Module {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: Table2Module,
            providers: []
        };
    }
}
