// @ericsson/oden/components/rule-container/rule-action/rule-action.component.ts
import { Component, HostBinding, Input, ViewEncapsulation, OnInit } from '@angular/core';
import { WhitelabelService } from '../../../core/services';
import { COLORWHITELABEL } from '../../../core/constants/color.whitelabel';
import { UnSub } from '../../../core/classes';
import { Subscription } from 'rxjs';

@Component({
    selector: 'cui-rule-action',
    templateUrl: 'rule-action.component.html',
    styleUrls: [ 'rule-action.component.scss' ],
    encapsulation: ViewEncapsulation.None
})
@UnSub()
export class RuleActionComponent implements OnInit {
    private _subscriptions: Subscription[] = [];
    @HostBinding('class.cui-rule-action') addClass: boolean = true;
    @Input() set color(color: string) {
        if ( color) {
            this._color = color;
        }
        this._getWhitelabel();
    }
    get color(): string {
        return this._color;
    }

    @Input() label: string;
    @Input() description?: string = '';

    private _color: string = '';
    constructor(private _whitelabelService: WhitelabelService) {
    }

    ngOnInit() {
        this._subscriptions.push(this._whitelabelService._whitelabelingLoaded.subscribe(() => {
            this._getWhitelabel();
        }));
        this._getWhitelabel();
    }

    private _getWhitelabel() {
        if (!this._color) {
            this._color = this._whitelabelService.getPropertyValue(COLORWHITELABEL.BRAND_FORTH);
        }
    }
}
