// @ericsson/oden/components/universal/universal.module.ts
import { ModuleWithProviders, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { IconModule } from '../icon/icon.module';
import { BackdropModule } from '../backdrop/backdrop.module';
import { UniversalTabsComponent } from './universal-tabs/universal-tabs.component';
import { UniversalAppsComponent } from './universal-apps/universal-apps.component';
import { UniversalOverlayComponent } from './universal-overlay/universal-overlay.component';
import { UniversalService } from './universal.service';
import { MenuCardModule } from '../menu-card/menu-card.module';
import { ContainerModule } from '../container/container.module';
import { FormModule } from '../form/form.module';
import { StickyModule } from '../sticky/sticky.module';
import { TabsModule } from '../tabs/tabs.module';
import { ListControllerModule } from '../list-controller/list-controller.module';
import { TranslateModule } from '@ngx-translate/core';
import { TooltipModule } from '../tooltip/tooltip.module';
import { ElementOverlayModule } from '../element-overlay/element-overlay.module';
import { GroupTitleModule } from '../group-title/group-title.module';

@NgModule({
    imports: [CommonModule, RouterModule, BackdropModule, IconModule, MenuCardModule,
        ContainerModule, FormModule, StickyModule, TabsModule, ListControllerModule, TranslateModule, TooltipModule, ElementOverlayModule,
        GroupTitleModule],
    exports: [UniversalTabsComponent, UniversalAppsComponent, UniversalOverlayComponent],
    declarations: [UniversalTabsComponent, UniversalAppsComponent, UniversalOverlayComponent],
    entryComponents: [UniversalAppsComponent]
})

export class UniversalModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: UniversalModule,
            providers: [UniversalService],
        };
    }
}
