// @ericsson/oden/components/main-menu/main-menu-route.model.ts
import { QuickAssign } from '../../core/helpers/quickassign.model';

export class BaseRoute extends QuickAssign {
    path: string; // Should be absolute path, optional for mainmenu route if expandableMenu is true.
    label: string;
    newWindow: boolean;
    externalApp: string; // Name of external application. Should be the same for all links of the same ext app.
    private _priority: number;

    get priority(): number {
        if (this._priority === undefined) {
            this.priority = Number.MAX_SAFE_INTEGER;
        }
        return this._priority;
    }

    set priority(value: number) {
        this._priority = value;
    }
}

export class ChildRoute extends BaseRoute {
    children?: Array<ChildRoute>;
}

export class ParentRoute extends BaseRoute {
    children: Array<ChildRoute>;
    mainMenuGroup: MainMenuRoute;
}

export class TopLevelRoute extends ParentRoute {
    applicationName: string;
}

export class MainMenuRoute extends BaseRoute {
    parentApp: string; // Should only be set if applicationName is null
    applicationName: string; // Should only be set if parentApp is null
    groupName: string; // Should only be set if object is route group
    parentGroup?: string; // Should only be set if groupName is null
    active?: boolean = false;
    configName?: string; // Should only be set if applicationName is null and should be unique
}
