// @ericsson/oden/components/table/table-row/table-row.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { ActionMenuModule } from '../../action-menu/action-menu.module';
import { ButtonModule } from '../../button/button.module';
import { FormModule } from '../../form/form.module';
import { IconModule } from '../../icon/icon.module';
import { TableRowComponent } from './table-row.component';
import { FormsModule } from '@angular/forms';
import { TableCellComponent } from './table-cell/table-cell.component';
import { TableCellDefaultComponent } from './table-cell/table-cell-default/table-cell-default.component';
import { StateBadgeComponent } from '../../state-badge/state-badge.component';
import { StateBadgeModule } from '../../state-badge/state-badge.module';
import { TableCellVersionComponent } from './table-cell/table-cell-version/table-cell-version.component';
import { TooltipModule } from '../../tooltip/tooltip.module';
import { TableCellEditComponent } from './table-cell/table-cell-edit/table-cell-edit.component';
import { TableCellEditDateComponent } from './table-cell/table-cell-edit/table-cell-edit-date/table-cell-edit-date.component';
import { TableCellEditTimeComponent } from './table-cell/table-cell-edit/table-cell-edit-time/table-cell-edit-time.component';
import { TableCellEditDateTimeComponent } from './table-cell/table-cell-edit/table-cell-edit-date-time/table-cell-edit-date-time.component';
import {
    TableCellEditMultiSelectComponent
} from './table-cell/table-cell-edit/table-cell-edit-multi-select/table-cell-edit-multi-select.component';
import {
    TableCellEditTextComponent
} from './table-cell/table-cell-edit/table-cell-edit-text/table-cell-edit-text.component';
import { TableCellEditTextAreaComponent } from './table-cell/table-cell-edit/table-cell-edit-text-area/table-cell-edit-text-area.component';
import { TableCellDateComponent } from './table-cell/table-cell-date/table-cell-date.component';
import { TableCellButtonComponent } from './table-cell/table-cell-button/table-cell-button.component';
import { TableRowPlaceholderComponent } from './table-row-placeholder/table-row-placeholder.component';
import {
    TableCellEditDateTimePeriodComponent
} from './table-cell/table-cell-edit/table-cell-edit-date-time-period/table-cell-edit-date-time-period.component';
import { TableCellCheckboxComponent } from './table-cell/table-cell-checkbox/table-cell-checkbox.component';
import { TableCellMetaComponent } from './table-cell/table-cell-meta/table-cell-meta.component';

@NgModule({
    imports: [CommonModule,
        ActionMenuModule,
        StateBadgeModule,
        ButtonModule,
        FormsModule,
        FormModule,
        IconModule,
        TooltipModule],
    exports: [TableRowComponent, TableRowPlaceholderComponent, TableCellMetaComponent],
    declarations: [
        TableRowComponent,
        TableCellComponent,
        TableCellDefaultComponent,
        TableCellVersionComponent,
        TableCellDateComponent,
        TableCellEditComponent,
        TableCellEditDateComponent,
        TableCellEditDateTimePeriodComponent,
        TableCellEditTimeComponent,
        TableCellEditDateTimeComponent,
        TableCellEditMultiSelectComponent,
        TableCellEditTextComponent,
        TableCellEditTextAreaComponent,
        TableCellButtonComponent,
        TableRowPlaceholderComponent,
        TableCellCheckboxComponent,
        TableCellMetaComponent
    ],
    entryComponents: [
        StateBadgeComponent,
        TableCellDefaultComponent,
        TableCellVersionComponent,
        TableCellDateComponent,
        TableCellEditComponent,
        TableCellEditDateComponent,
        TableCellEditTimeComponent,
        TableCellEditDateTimeComponent,
        TableCellEditMultiSelectComponent,
        TableCellEditTextComponent,
        TableCellEditDateTimePeriodComponent,
        TableCellCheckboxComponent,
        TableCellEditTextAreaComponent,
        TableCellButtonComponent
    ]
})

export class TableRowModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: TableRowModule,
            providers: []
        };
    }
}
