// @ericsson/oden/components/guided-workflow/guided-workflow.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';

import { ActionMenuModule } from '../action-menu/action-menu.module';
import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import { ButtonModule } from '../button/button.module';
import { ContainerModule } from '../container/container.module';
import { FormModule } from '../form/form.module';
import { IconModule } from '../icon/icon.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { GuidedWorkflowComponent } from './guided-workflow.component';
import { GuidedWorkflowService } from './guided-workflow.service';
import { LoaderModule } from '../loader/loader.module';

@NgModule({
    imports: [CommonModule, ContainerModule, FormModule, ButtonModule, LoaderModule,
        BreadcrumbModule, IconModule, ActionMenuModule, TooltipModule, TranslateModule.forChild()],
    exports: [GuidedWorkflowComponent],
    declarations: [GuidedWorkflowComponent],
    entryComponents: [GuidedWorkflowComponent]
})

export class GuidedWorkflowModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: GuidedWorkflowModule,
            providers: [GuidedWorkflowService]
        };
    }
}
