// @ericsson/oden/components/block-tree/block-tree.module.ts
import { BlockTreeService } from './block-tree.service';
import { BlockTreeComponent } from './block-tree.component';
import { BlockTreeBranchComponent } from './branch/branch.component';
import { BlockTreeBranchLeafComponent } from './branch/leaf/leaf.component';
import { CommonModule } from '@angular/common';
import { ActionMenuModule } from '../action-menu/action-menu.module';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { IconModule } from '../icon';
import { DragAndDropModule } from '../drag-and-drop';
import { TagsModule } from '../tags';
import { FormModule } from '../form';
import { FormsModule } from '@angular/forms';
import { TactileBlockModule } from '../tactile-block';
import { PipesModule } from '../../core/pipes/pipes.module';
import { TooltipModule } from '../tooltip';
import { ButtonModule } from '../button/button.module';
import { BlockDropElementDirective } from './block-dropelement.directive';

@NgModule({
    imports: [CommonModule, RouterModule, IconModule, ActionMenuModule, DragAndDropModule, TooltipModule,
        TagsModule, FormModule, FormsModule, TactileBlockModule, PipesModule, TooltipModule, ButtonModule],
    exports: [BlockTreeComponent, BlockTreeBranchComponent, BlockTreeBranchLeafComponent, BlockDropElementDirective],
    declarations: [
        BlockTreeComponent, BlockTreeBranchComponent, BlockTreeBranchLeafComponent, BlockDropElementDirective
    ]
})

export class BlockTreeModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: BlockTreeModule,
            providers: [ BlockTreeService ]
        };
    }
}
