// @ericsson/oden/components/oden-form/oden-form.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
import { GestureService } from '../../core/services/gesture.service';
import { TooltipModule } from '../tooltip/tooltip.module';
import { OdenFormService } from '../../core/services/oden-form.service';
import { IconModule } from '../icon/icon.module';
import { TranslateModule } from '@ngx-translate/core';
import { OdenCheckboxDirective } from './oden-checkbox/oden-checkbox.directive';
import { OdenComboboxComponent } from './oden-combobox/oden-combobox.component';
import { OdenErrorComponent } from './oden-error/oden-error.component';
import { OdenFormFieldComponent } from './oden-form-field/oden-form-field.component';
import { OdenFormFieldsetComponent } from './oden-form-fieldset/oden-form-fieldset.component';
import { OdenFormGroupComponent } from './oden-form-group/oden-form-group.component';
import { OdenFormHeaderDirective } from './oden-form-header/oden-form-header.directive';
import { OdenInputDirective } from './oden-input/oden-input.directive';
import { OdenIssueBaseComponent } from './oden-form-base-classes/oden-issue-base-component';
import { OdenLabelDirective } from './oden-label/oden-label.directive';
import { OdenOptionComponent } from './oden-combobox/oden-option/oden-option.component';
import { OdenRadioDirective } from './oden-radio/oden-radio.directive';
import { OdenSelectionInputBaseDirective } from './oden-form-base-classes/oden-selection-input-base-directive';
import { OdenSwitchDirective } from './oden-switch/oden-switch.directive';
import { OdenTextareaDirective } from './oden-textarea/oden-textarea.directive';
import { OdenTextInputBaseDirective } from './oden-form-base-classes/oden-text-input-base-directive';
import { OdenWarningDirective } from './oden-warning/oden-warning.directive';

@NgModule({
    imports: [
        CommonModule,
        TooltipModule,
        IconModule,
        TranslateModule.forChild()
    ],
    exports: [
        OdenCheckboxDirective,
        OdenComboboxComponent,
        OdenErrorComponent,
        OdenFormFieldComponent,
        OdenFormFieldsetComponent,
        OdenFormGroupComponent,
        OdenFormHeaderDirective,
        OdenInputDirective,
        OdenLabelDirective,
        OdenOptionComponent,
        OdenRadioDirective,
        OdenSwitchDirective,
        OdenTextareaDirective,
        OdenWarningDirective
    ],
    declarations: [
        OdenCheckboxDirective,
        OdenComboboxComponent,
        OdenErrorComponent,
        OdenFormFieldComponent,
        OdenFormFieldsetComponent,
        OdenFormGroupComponent,
        OdenFormHeaderDirective,
        OdenInputDirective,
        OdenIssueBaseComponent,
        OdenLabelDirective,
        OdenOptionComponent,
        OdenRadioDirective,
        OdenSelectionInputBaseDirective,
        OdenSwitchDirective,
        OdenTextareaDirective,
        OdenTextInputBaseDirective,
        OdenWarningDirective
    ],
    providers: [
        OdenFormService
    ]
})

export class OdenFormModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: OdenFormModule,
            providers: [
                { provide: HAMMER_GESTURE_CONFIG, useClass: GestureService }
            ]
        };
    }
}
