// @ericsson/oden/core/models/library-group.model.ts
import { TactileBlockItem } from './tactile-block-item.model';
import { ListControllerCategoryFilterGroup } from './list-controller-category-filter-group.model';
import { UUID } from '../helpers/uuid.class';

export class LibraryGroup {
    id: string = 'library-group-' + UUID.create();
    title: string;
    items: Array<TactileBlockItem> = [];
    active: boolean = false;
    categoryFilters?: Array<ListControllerCategoryFilterGroup> = [];
    filterString?: string = '';
    showNumberOfResults?: boolean = false;
    isFiltersExpanded?: boolean = false;
    expandableTactileBlocks?: boolean = false;

    constructor(obj: {
        id?: string,
        title?: string,
        items?: Array<TactileBlockItem>,
        active?: boolean,
        categoryFilters?: Array<ListControllerCategoryFilterGroup>;
        filterString?: string;
        showNumberOfResults?: boolean;
        isFiltersExpanded?: boolean;
        expandableTactileBlocks?: boolean;
    } = {}) {
        Object.assign(this, obj);
    }
}
