// @ericsson/oden/core/models/table-row-data.model.ts
import { QuickAssign } from '../helpers/quickassign.model';
import { ITableRowData } from '../interfaces/table-row-data.interface';
import { UUID } from '../helpers/uuid.class';
import { ActionItem } from './action-item.model';

export class TableRowData extends QuickAssign implements ITableRowData {
    /** @deprecated - use Templates instead! */
    edit: boolean;
    /** @deprecated - use Templates instead! */
    focus: boolean;
    /** @deprecated - use Templates instead! */
    editError: {} = {};
    hidden: boolean = false;
    selected: boolean = false;
    /** @deprecated - use Templates instead! */
    path: string;
    private _uuid: string;
    singleActions: boolean = true;
    /** @deprecated - use Templates instead! */
    editDisabled: Array<string>;
    /** @deprecated - use Templates instead! */
    editedFields: {} = {};
    multiRow?: boolean;
    multiRowWrap?: boolean;
    rowActions?: Array<ActionItem>;
    get uuid(): string {
        if (!this._uuid) {
            this._uuid = UUID.create();
        }
        return this._uuid;
    }

    set uuid(val: string) {
        this._uuid = val;
    }
}
