// @ericsson/oden/core/models/workflow-guide.model.ts
import { WorkflowGuideStep } from './workflow-guide-step.model';
import { BreadcrumbItem } from './breadcrumb.item';
import { ActionItem } from './action-item.model';
import { GUIDED_WORKFLOW } from '../constants/guided-workflow.constants';
export class WorkflowGuide {
    id: string;
    steps: Array<WorkflowGuideStep>;
    overviewSteps?: Array<WorkflowGuideStep>;
    activeStep?: number = 1;
    visible?: boolean = false;
    minimized?: boolean = false;
    maximized?: boolean = false;
    breadcrumb?: Array<BreadcrumbItem>;
    actionMenuActions?: Array<ActionItem>;
    actionMenuEvent?: any;
    finishTitle: string;
    finishLocked: boolean = true;
    hideBack: boolean = false;
    finishEvent?: any;
    pauseEvent?: any;
    goToEvent?: any;
    maximizeEvent?: any;
    position?: string = GUIDED_WORKFLOW.TOP;
    data?: any;

    constructor(obj: {
        id?: string;
        steps?: Array<WorkflowGuideStep>;
        overviewSteps?: Array<WorkflowGuideStep>;
        overviewLoading?: boolean;
        activeStep?: number;
        visible?: boolean;
        minimized?: boolean;
        maximized?: boolean;
        breadcrumb?: Array<BreadcrumbItem>;
        actionMenuActions?: Array<ActionItem>;
        actionMenuEvent?: any;
        finishTitle?: string;
        finishLocked?: boolean;
        hideBack?: boolean;
        finishEvent?: any;
        pauseEvent?: any;
        goToEvent?: any;
        maximizeEvent?: any;
        position?: string;
        data?: any;
    } = {}) {
        Object.assign(this, obj);
    }
}
