// @ericsson/oden/core/models/workflow-guide-step.model.ts
export class WorkflowGuideStep {
    number: number;
    title: string;
    description: string;
    path: string; // browser will go to this path when the step is accessed
    locked: boolean = false; // a locked step cannot be accessed
    lockedMessage: string = ''; // this message will show up in tooltip over the locked button
    dirty?: boolean = false; // useful attribute to store if things have changed in the step
    hideFutureSteps?: boolean = false; // set this to true to temporary hide title and description for upcoming steps
    data?: any; // save custom data here
    nextDisabled?: boolean = false; // set this to true disable the Next button. Only applicable if hideFutureSteps is true.
    nextDisabledMessage: string = ''; // this message will show up in tooltip over the disabled Next button (for hideFutureSteps only)

    constructor(obj: {
        number?: number;
        title?: string;
        description?: string;
        path?: string;
        locked?: boolean;
        lockedMessage?: string;
        dirty?: boolean;
        hideFutureSteps?: boolean;
        data?: any;
        nextDisabled?: boolean;
        nextDisabledMessage?: string;
    } = {}) {
        Object.assign(this, obj);
    }
}
