// @ericsson/oden/components/validation-summary/validation-summary.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';

import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import { ButtonModule } from '../button/button.module';
import { FormModule } from '../form/form.module';
import { IconModule } from '../icon/icon.module';
import { StickyModule } from '../sticky/sticky.module';
import { TabsModule } from '../tabs/tabs.module';
import { ValidationSummaryComponent } from './validation-summary.component';
import { TooltipModule } from '../tooltip/tooltip.module';
import { SeeAllDialogComponent } from './see-all-dialog/see-all-dialog.component';


@NgModule({
    imports: [
        BreadcrumbModule,
        ButtonModule,
        CommonModule,
        TabsModule,
        StickyModule,
        IconModule,
        FormModule,
        FormsModule,
        TranslateModule.forChild(),
        TooltipModule
    ],
    exports: [
        ValidationSummaryComponent
    ],
    declarations: [
        SeeAllDialogComponent,
        ValidationSummaryComponent
    ],
    entryComponents: [
        SeeAllDialogComponent
    ]
})

export class ValidationSummaryModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: ValidationSummaryModule
        };
    }
}
