// @ericsson/oden/components/error-page/error-page.component.ts
import { Component, ViewEncapsulation, Input, HostBinding, EventEmitter, TemplateRef, ViewChild } from '@angular/core';
import { ButtonActionItem } from '../../core/models/button-action-item.model';

@Component({
    selector: 'cui-error-page',
    templateUrl: 'error-page.component.html',
    styleUrls: ['error-page.component.scss'],
    encapsulation: ViewEncapsulation.None,
})
export class ErrorPageComponent {
    @HostBinding('class.cui-error-page') addClass: boolean = true;
    @HostBinding('class.full') @Input() fullSize: boolean = false;
    @ViewChild('defaultTemplate', {read: TemplateRef}) defaultTemplate;

    onAction: EventEmitter<ButtonActionItem> = new EventEmitter<ButtonActionItem>();

    @Input() title: string;
    @Input() message: string;
    @Input() messageTemplate: TemplateRef<any>;
    @Input() statusCode: number;
    @Input() actions?: Array<ButtonActionItem> = [];

    actionClicked(item: ButtonActionItem): void {
        this.onAction.emit(item);
    }

    get template(): TemplateRef<any> {
        return this.messageTemplate || this.defaultTemplate;
    }
}
