// @ericsson/cus-ui/boilerplate/app/shared/profile.service.mock.ts
import { Injectable } from '@angular/core';
import { DialogConfig,
    ModalHelperService,
    DIALOG_TYPE,
    DialogComponent } from '@ericsson/oden/modules';
import { ProfileModel, ModalActionItem } from '@ericsson/oden/core';

/**
 * Service for setting user data in system bar
 */
@Injectable()
export class ProfileMockService {
    dialog: DialogComponent;

    constructor(private _modalHelperService: ModalHelperService) {}

    /**
     * Returns an ProfileModel of mocked data.
     *
     * @returns ProfileModel
     */
    public getProfileData(): ProfileModel {
        return new ProfileModel(
            {
                name: 'User',
                imgUrl: 'assets/user.png'
            });
    }

    /**
     * Opens a system alert triggered from view.
     */
    public openLogoutAlert(): void {
        let config: DialogConfig = new DialogConfig({
            message: `
                <p>Are you sure you want to Logout?</p>
            `,
            actions: [
                new ModalActionItem({
                    label: 'Cancel',
                    action: 'cancel',
                    cssClass: 'discard'
                }),
                new ModalActionItem({
                    label: 'Ok',
                    action: 'ok',
                    cssClass: 'primary'
                })
            ],
            type: DIALOG_TYPE.SYSTEM_ALERT
        });

        this.dialog = this._modalHelperService.createDialog(config);
        this.dialog.onAction.subscribe((actionItem) => {
            switch (actionItem.action) {
                case 'ok':
                    this.dialog.close();
                    break;
                case 'cancel':
                    this.dialog.close();
                    break;
                default:
            }
        });
    }
}
