// @ericsson/oden/components/container/row/column/column.component.ts
import { Component, ElementRef, Input, OnChanges, Renderer2, SimpleChanges, ViewEncapsulation } from '@angular/core';
import { COLUMN_ALIGNMENT } from '../../../../core/constants/column-alignment.constants';

@Component({
    selector: 'cui-column',
    templateUrl: 'column.component.html',
    styleUrls: ['column.component.scss'],
    encapsulation: ViewEncapsulation.None,
})

export class ColumnComponent implements OnChanges {
    @Input() align: string | Array<string> = '';

    constructor(private _element: ElementRef, private _renderer: Renderer2) { }

    ngOnChanges(changes: SimpleChanges): void {
        if (changes['align']) {
            let align = this.align;
            if (typeof this.align === 'string') {
                align = [<string>this.align];
            }
            this._renderer.removeClass(this._element.nativeElement, 'center-xs');
            this._renderer.removeClass(this._element.nativeElement, 'end-xs');
            this._renderer.removeClass(this._element.nativeElement, 'bottom-xs');
            this._renderer.removeClass(this._element.nativeElement, 'middle-xs');

            (<Array<string>>align).forEach((_align: string) => {
                switch (_align) {
                    case COLUMN_ALIGNMENT.END:
                        this._renderer.addClass(this._element.nativeElement, 'end-xs');
                        break;
                    case COLUMN_ALIGNMENT.BOTTOM:
                        this._renderer.addClass(this._element.nativeElement, 'bottom-xs');
                        break;
                    case COLUMN_ALIGNMENT.CENTER:
                        this._renderer.addClass(this._element.nativeElement, 'center-xs');
                        break;
                    case COLUMN_ALIGNMENT.MIDDLE:
                        this._renderer.addClass(this._element.nativeElement, 'middle-xs');
                        break;
                }
            });
        }
    }
}
