// @ericsson/oden/components/characteristics/characteristic-container/characteristic-container.component.ts
import {
  Component, Input, OnChanges
} from '@angular/core';
import { Characteristic, CharacteristicValue } from '../../../core/models';
import { CharacteristicHelperService } from '../characteristic.helper.service';
import { CharacteristicService } from './characteristic.service';
import { CharacteristicValueType } from '../../../core/constants';

@Component({
  selector: 'cui-characteristic-container',
  templateUrl: 'characteristic-container.component.html',
  providers: [CharacteristicService]
})
export class CharacteristicContainerComponent implements OnChanges {

  @Input('characteristic')
  get characteristic(): Characteristic {
    return this._characteristicService.characteristic;
  }
  set characteristic(value: Characteristic) {
    this._characteristicService.characteristic = value;
  }

  title: string;
  data: CharacteristicValue[] = [];
  valueType: CharacteristicValueType;
  valueTypes = CharacteristicValueType;

  constructor(
    private _characteristicService: CharacteristicService,
    private _characteristicHelper: CharacteristicHelperService) { }

  ngOnChanges() {
    this.title = this.characteristic.name;
    this.data = this._characteristicHelper.getCharacteristicValue(this.characteristic);
    this.valueType = this.characteristic.valueType;
  }

}
