// @ericsson/oden/components/table/table-row/table-cell/table-cell-button/table-cell-button.component.ts
import { Component, EventEmitter, Input, Output, ViewEncapsulation, HostBinding } from '@angular/core';
import { FIELD_EVENT_TYPE, TableEditEvent, TableFieldEvent } from '../../../../../core/models/table-field.event';
import { TableButtonAction } from '../../../../../core/models/table-button-action.model';

@Component({
    selector: 'cui-table-cell-button',
    templateUrl: './table-cell-button.component.html',
    styleUrls: ['./table-cell-button.component.scss'],
    encapsulation: ViewEncapsulation.None
})

export class TableCellButtonComponent {
    @Input() data: TableButtonAction;
    @Output() onFieldEvent: EventEmitter<TableEditEvent> = new EventEmitter<TableEditEvent>();
    @HostBinding('class.cui-table-cell-button') addClass: boolean = true;

    constructor() {
    }

    /**
     * Emits onFieldChange event when a field is changed
     * @param value
     * @param element
     */
    onClick(element: Element) {
        this.onFieldEvent.emit(new TableFieldEvent({
                eventType: FIELD_EVENT_TYPE.CLICK,
                element: element,
                value: this.data
            }
        ));
    }
}
