// @ericsson/oden/components/library/library.module.ts
import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';

import { BlockModule } from '../block/block.module';
import { ButtonModule } from '../button/button.module';
import { ContainerModule } from '../container/container.module';
import { DraggableModule } from '../draggable/draggable.module';
import { FormModule } from '../form/form.module';
import { ListControllerModule } from '../list-controller/list-controller.module';
import { TabsModule } from '../tabs/tabs.module';
import { TactileBlockModule } from '../tactile-block/tactile-block.module';
import { TooltipModule } from '../tooltip/tooltip.module';
import { LibraryService } from './../../core/services/library.service';
import { LibraryGroupComponent } from './library-group/library-group.component';
import { LibraryComponent } from './library.component';
import { PaginationModule } from '../pagination/pagination.module';
import { ElementOverlayModule } from '../element-overlay/element-overlay.module';
import { DragAndDropModule } from '../drag-and-drop/drag-and-drop.module';
import { ResizeService } from '../../core/services/resize.service';
import { SelectReceiverDialogComponent } from './select-receiver-dialog/select-receiver-dialog.component';

@NgModule({
    imports: [CommonModule, FormsModule, FormModule, ContainerModule, ButtonModule, PaginationModule, DragAndDropModule,
        BlockModule, TabsModule, ListControllerModule, TactileBlockModule, DraggableModule, TooltipModule, ElementOverlayModule,
        TranslateModule.forChild()],
    exports: [LibraryComponent, LibraryGroupComponent],
    declarations: [LibraryComponent, LibraryGroupComponent, SelectReceiverDialogComponent],
    providers: [ResizeService],
    entryComponents: [SelectReceiverDialogComponent]
})

export class LibraryModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: LibraryModule,
            providers: [LibraryService]
        };
    }
}
